var documenterSearchIndex = {"docs":
[{"location":"examples/GPU/#TraitSimulation-explores-GPU-computing","page":"Exploring GPU Potential","title":"TraitSimulation explores GPU computing","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Author: Sarah Ji","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"In this notebook, we will explore some GPU capabilities of the Julia language in the CuArrays package, and present the potential for users to extend the functions in TraitSimulation.jl to accomodate advances in GPU computing. Although GPU computing is still under active development, it is fruitful for researchers to be aware of the most modern computing technologies available to users through open source options. ","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"We will first present the major bottle necks of the simulation algorithm and benchmark the optimized code (using BLAS) on CPU vs. on GPU. Then we will simulate the symmetric Bivariate trait using TraitSimulation on CPU first, then GPU for comparison. ","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"First the simulation algorithm first performs the Cholesky factors of the variance/covariance matrices and stores them using the @vc macro.\nThen we transform the standard normal matrix using the Cholesky factors above","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Note useres with many variance components or extremely large number of subjects may benefit most from exploring GPU.  Additionally, since the default setting in the notebook is 1 thread, so I will only compare the unthreaded CPU code vs. the GPU code. ","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"versioninfo()","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Julia Version 1.4.0\nCommit b8e9a9ecc6 (2020-03-21 16:36 UTC)\nPlatform Info:\n  OS: Linux (x86_64-pc-linux-gnu)\n  CPU: Intel(R) Core(TM) i9-9920X CPU @ 3.50GHz\n  WORD_SIZE: 64\n  LIBM: libopenlibm\n  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)","category":"page"},{"location":"examples/GPU/#Check-available-devices-on-this-machine-and-show-their-capability","page":"Exploring GPU Potential","title":"Check available devices on this machine and show their capability","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"using CuArrays, CUDAdrv\nusing CuArrays.CURAND\n\nfor device in CuArrays.devices()\n    @show capability(device)\nend","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"capability(device) = v\"7.5.0\"","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"We use simulated data and include the exploratory TraitSimulation on GPU code, in the file exploring_gpu_simulation.jl. We encourage users with the right NVIDIA GPU machine to explore these options for themselves","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"include(\"/home/sarahji/TraitSimulation.jl/src/exploring_gpu_simulation.jl\");","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"using GLM, LinearAlgebra, Random, BenchmarkTools\nusing SnpArrays, Statistics\n\nRandom.seed!(1234)\n\nfunction generateSPDmatrix(n)\n    A = rand(n)\n    m = 0.5 * (A * A')\n    PDmat = m + (n * Diagonal(ones(n)))\nend\n\n\nfunction generateRandomVCM(n::Int64, p::Int64, d::Int64, m::Int64)\n    # n-by-p design matrix\n    X = randn(n, p)\n\n    # p-by-d mean component regression coefficient for each trait\n    B = rand(p, d)\n\n    V = ntuple(x -> zeros(n, n), m)\n    for i = 1:m-1\n      copy!(V[i], generateSPDmatrix(n))\n    end\n    copy!(V[end], Diagonal(ones(n))) # last covarianec matrix is identity\n\n    # a tuple of m d-by-d variance component parameters\n    Î£ = ntuple(x -> zeros(d, d), m)\n    for i in 1:m\n      copy!(Î£[i], generateSPDmatrix(d))\n    end\n    return(X, B, Î£, V)\nend\n\n\nn = 5000 # number of people\np = 3   # number of fixed effects\nd = 10   # number of traits\nm = 10   # number of variance components\n\nX, Î², Î£, V = generateRandomVCM(n, p, d, m);","category":"page"},{"location":"examples/GPU/#.-Cholesky-Decomposition-of-Variance-Components","page":"Exploring GPU Potential","title":"1. Cholesky Decomposition of Variance Components","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"The major bottleneck in the simulation process is the cholesky decomposition of the variance covariance matrices (variance components). We note that if users have many variance components will benefit from having the GPU option available to their software. ","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"First we compute the Cholesky Decomposition of a single nxn matrix, V[1], then we use the @vc macro and make a GPU alternative to compare.","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"We see that for a 5000 by 5000 matrix, the GPU code is more than 30x faster than the cholesky decomposition computed on CPU.","category":"page"},{"location":"examples/GPU/#Cholesky-on-CPU-for-a-single-variance-component-(n-x-n)","page":"Exploring GPU Potential","title":"Cholesky on CPU for a single variance component (n x n)","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"# Cholesky on CPU for a single trait\n@show size(V[1])\n@benchmark cholesky((Symmetric($V[1])))","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"size(V[1]) = (5000, 5000)\n\nBenchmarkTools.Trial: \n  memory estimate:  190.74 MiB\n  allocs estimate:  6\n  --------------\n  minimum time:     277.191 ms (0.00% GC)\n  median time:      292.412 ms (0.00% GC)\n  mean time:        326.911 ms (1.62% GC)\n  maximum time:     769.881 ms (4.69% GC)\n  --------------\n  samples:          16\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/#Cholesky-on-GPU-for-a-single-variance-component-(n-x-n)","page":"Exploring GPU Potential","title":"Cholesky on GPU for a single variance component (n x n)","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"@time V1_gpu = CuArray{Float32}(V[1]) # Move the Variance Covariance Matrix V from CPU to GPU environment\n@benchmark cholesky((Symmetric($V1_gpu)))","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"  0.959381 seconds (1.94 M allocations: 191.373 MiB)\n\nBenchmarkTools.Trial: \n  memory estimate:  4.00 KiB\n  allocs estimate:  118\n  --------------\n  minimum time:     10.111 ms (0.00% GC)\n  median time:      10.289 ms (0.00% GC)\n  mean time:        10.339 ms (0.34% GC)\n  maximum time:     14.302 ms (19.20% GC)\n  --------------\n  samples:          484\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"We made this @vc macro to provide a more flexible input option for users, but we see that there is a cost for comfort!","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"The single cholesky is about 30 fold speed up but using the macro its only about 6x  So we note that for maximum efficiency, when users have large datasets, they can bypass this and just store the cholesky factors necessary to perform the transformation.","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"vc_cpu = @vc Î£[1] âŠ— V[1] + Î£[2] âŠ— V[2] # make the vc object with just 2 variance components\ntrait = VCMTrait(X, Î², [Î£...], [V...]) # make the trait object with all m = 10 variance components.\n\n@benchmark vc_cpu = @vc $Î£[1] âŠ— $V[1] + $Î£[2] âŠ— $V[2]","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  381.47 MiB\n  allocs estimate:  25\n  --------------\n  minimum time:     1.609 s (0.31% GC)\n  median time:      1.801 s (3.48% GC)\n  mean time:        2.100 s (3.78% GC)\n  maximum time:     2.890 s (5.90% GC)\n  --------------\n  samples:          3\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"vc_gpu = @vc_gpu Î£[1] âŠ— V[1] + Î£[2] âŠ— V[2]\n@benchmark vc_gpu = @vc_gpu $Î£[1] âŠ— $V[1] + $Î£[2] âŠ— $V[2]","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  16.78 KiB\n  allocs estimate:  497\n  --------------\n  minimum time:     274.375 ms (0.00% GC)\n  median time:      277.834 ms (0.00% GC)\n  mean time:        292.647 ms (0.22% GC)\n  maximum time:     350.880 ms (0.83% GC)\n  --------------\n  samples:          18\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/#.-Matrix-Normal-Transformation-by-Cholesky-Factors","page":"Exploring GPU Potential","title":"2. Matrix Normal Transformation by Cholesky Factors","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Each variance component will require it's own simulation from the standard normal distribution and the appropriate transformations by the cholesky factors of the row and column variances. ","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"We saw separately that the cholesky decomposition is sped up by an order of magnitude on GPU vs. on CPU in step 1. Now, to complete the simulation we will perform step 2 to multiply on the left and right to transform the simulated trait to have the desired covariance.","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"There is a 30x fold speed up in JUST transforming the matrix normal using GPU.","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"function cpu_single_vc(trait, CholÎ£, CholV)\n    randn!(trait.Z)\n    BLAS.trmm!('L', 'U', 'T', 'N', 1.0, CholV, trait.Z)\n    BLAS.trmm!('R', 'U', 'N', 'N', 1.0, CholÎ£, trait.Z)\n    trait.Z\nend\n\nCholÎ£ = trait.vc[1].CholÎ£ # grab the cholesky factors of Î£ (dxd) from step 1\nCholV = trait.vc[1].CholV # grab the cholesky factor of V (nxn) from step 1\n\n@benchmark cpu_single_vc($trait, $CholÎ£, $CholV)","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  0 bytes\n  allocs estimate:  0\n  --------------\n  minimum time:     19.655 ms (0.00% GC)\n  median time:      20.967 ms (0.00% GC)\n  mean time:        21.233 ms (0.00% GC)\n  maximum time:     23.723 ms (0.00% GC)\n  --------------\n  samples:          236\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"function gpu_single_vc(Z_d, CholÎ£, CholV)\n    randn!(Z_d)\n    CuArrays.CUBLAS.trmm!('L', 'U', 'T', 'N', 1.0, CholV, Z_d, similar(Z_d))\n    CuArrays.CUBLAS.trmm!('R', 'U', 'N', 'N', 1.0, CholÎ£, Z_d, similar(Z_d))\n    Z_d\nend\n\nZ_d = CuArray{Float64}(trait.Z)\nCholÎ£_gpu = vc_gpu[1].CholÎ£ # grab the cholesky factors of Î£ (dxd) from step 1\nCholV_gpu = vc_gpu[1].CholV # grab the cholesky factor of V (nxn) from step 1\n\n@benchmark gpu_single_vc($Z_d, $CholÎ£_gpu, $CholV_gpu)","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  784 bytes\n  allocs estimate:  29\n  --------------\n  minimum time:     16.213 Î¼s (0.00% GC)\n  median time:      1.061 ms (0.00% GC)\n  mean time:        1.010 ms (0.08% GC)\n  maximum time:     10.271 ms (38.50% GC)\n  --------------\n  samples:          4946\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/#Perform-TraitSimulation","page":"Exploring GPU Potential","title":"Perform TraitSimulation","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Here we just compare a single simulation on CPU vs. on GPU.  This is the function that is continuously called in the power calculation. It writes over the field Y, the simulated results after aggregating the simulation results of the m variance components. (random effects) ","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"First we see the simulation for a single variance component\nThen we see the simulation for the entire variance component set in trait.vc","category":"page"},{"location":"examples/GPU/#TraitSimulation-for-a-Single-Variance-Component","page":"Exploring GPU Potential","title":"TraitSimulation for a Single Variance Component","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"We see this is roughly 20 times faster on GPU than on CPU for a single V_n times n Sigma_d times d","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"@benchmark TraitSimulation.simulate_matrix_normal!($trait.Z, $vc_cpu[1])","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  0 bytes\n  allocs estimate:  0\n  --------------\n  minimum time:     27.357 ms (0.00% GC)\n  median time:      32.112 ms (0.00% GC)\n  mean time:        40.311 ms (0.00% GC)\n  maximum time:     83.568 ms (0.00% GC)\n  --------------\n  samples:          124\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Z_d = CuArray{Float64}(trait.Z)\n@benchmark simulate_matrix_normal_gpu!($Z_d, $vc_gpu[1])","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  784 bytes\n  allocs estimate:  29\n  --------------\n  minimum time:     17.176 Î¼s (0.00% GC)\n  median time:      1.060 ms (0.00% GC)\n  mean time:        1.009 ms (0.00% GC)\n  maximum time:     1.468 ms (0.00% GC)\n  --------------\n  samples:          4950\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/#TraitSimulation-for-the-set-of-m-Variance-Components","page":"Exploring GPU Potential","title":"TraitSimulation for the set of m Variance Components","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"We see above that the simulation for just a single variance component is roughly 30 times faster. ","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Since we have m = 10 variance components in our model, we see that in this case the simulation on GPU is 100x faster.","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Y = zero(trait.Î¼)\n@benchmark TraitSimulation.VCM_trait_simulation($Y, $trait.Z, $trait.Î¼, $trait.vc)","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  0 bytes\n  allocs estimate:  0\n  --------------\n  minimum time:     200.812 ms (0.00% GC)\n  median time:      257.338 ms (0.00% GC)\n  mean time:        266.100 ms (0.00% GC)\n  maximum time:     626.675 ms (0.00% GC)\n  --------------\n  samples:          19\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"Y_d = zeros(CuArray{Float32}, size(trait.Z))\nÎ¼_d = zeros(CuArray{Float32}, size(trait.Î¼))\n@benchmark VCM_trait_simulation_gpu($Y_d, $Î¼_d, $Z_d, $vc_gpu)","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"BenchmarkTools.Trial: \n  memory estimate:  10.67 KiB\n  allocs estimate:  283\n  --------------\n  minimum time:     67.968 Î¼s (0.00% GC)\n  median time:      2.134 ms (0.00% GC)\n  mean time:        2.056 ms (0.00% GC)\n  maximum time:     2.361 ms (0.00% GC)\n  --------------\n  samples:          2430\n  evals/sample:     1","category":"page"},{"location":"examples/GPU/#Threading:","page":"Exploring GPU Potential","title":"Threading:","text":"","category":"section"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"For users who wish to do multiple simulation runs simultaneously, we recommend to set the machinery to use threading. Users can check using the command: Threads.nthreads() to ensure multi-threading is on. TraitSimulation will automatically use the Threading option for multiple TraitSimulation. To set the number of threads, users should follow the documentation on Threads.jl and ensure before starting Julia to specify the desired number of threads using export JULIA_NUM_THREADS=4.","category":"page"},{"location":"examples/GPU/","page":"Exploring GPU Potential","title":"Exploring GPU Potential","text":"For users who are using the Threading option and are seeing some variation in the benchmarking results, make sure that the julia number of threads and the BLAS number of threads are not confusing one another and specify the command: LinearAlgebra.BLAS.set_num_threads(1) to make simulations more consistent.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Trait-Simulation-Ordered-Multinomial-Power-(UKBiobank)","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"Authors: Sarah Ji, Janet Sinsheimer, Kenneth Lange, Hua Zhou","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"In this notebook we show how to use the TraitSimulation.jl package we illustrate how TraitSimulation.jl can easily simulate traits from genotype data, all within the OpenMendel universe. Operating within this universe brings potential advantages over the available software(s) when needed for downstream analysis or study design. Using just a few calls on the command line to the appropriate packages within the OpenMendel, we demonstrate how to use the OpenMendel code pipeline to construct the genetic model, simulate the trait and find the power for the Ordinal Multinomial Trait. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Background","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Background","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"The ordinal multinomial model is a powerful way to model stages of disease progression. For a disease like diabetes, the monitoring of the disease's progression may be improved based on several different biomarkers detected in the EHR data, such as diabetes diagnostic codes, diabetes medication, hyperglycemia in blood results defined by HbA1c and fasting glucose levels, and presence of diabetes process of care codes. Then these algorithm can better categorize individuals into disease progression categories that relate to how likely they are to have diabetes. This is motivated by the need for early diagnoses and effective intervention techniques. For a detailed review on the published ordinal model, we provide in the references our most recent published paper in Genetic Epidemiology, Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"This software package, TraitSimuliation.jl addresses the need for simulated trait data in genetic analyses.  This package generates data sets that will allow researchers to accurately check the validity of programs and to calculate power for their proposed studies. This package gives users the ability to easily simulate phenotypic traits under generalized linear models (GLMs) or variance component models (VCMs) conditional on PLINK formatted genotype data [3]. In addition, we demo customized simulation utilities that accompany specific genetic analysis options in Open-Mendel; for example, ordered, multinomial traits. We demonstrate these simulation utilities on the example dataset described below.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Demonstration","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Demonstration","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Example-Data","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Example Data","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"The data in this demo is restricted to a subset of individuals of European ancestry excluding first- and second-degree relatives based off of kinship estimates. We filtered samples by 98% genotyping success rate on all chromosomes and SNPs by 99% genotyping success rate. These measures result in n = 185,565 individuals and 470,228 SNPs for analysis. A more detailed description of the data can be found in the references 3 The data from this notebook can be found in the UK Biobank repository under Project ID: 48152 and 15678 at https://www.ukbiobank.ac.uk with the permission of UK Biobank. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"We use the OpenMendel package SnpArrays.jl to both read in and write out PLINK formatted files. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Double-check-that-you-are-using-Julia-version-1.0-or-higher-by-checking-the-machine-information","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Double check that you are using Julia version 1.0 or higher by checking the machine information","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"versioninfo()","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"Julia Version 1.3.0\nCommit 46ce4d7933 (2019-11-26 06:09 UTC)\nPlatform Info:\n  OS: Linux (x86_64-pc-linux-gnu)\n  CPU: Intel(R) Core(TM) i9-9920X CPU @ 3.50GHz\n  WORD_SIZE: 64\n  LIBM: libopenlibm\n  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"using Plots, DataFrames, LinearAlgebra, StatsFuns, CSV\nusing SnpArrays, TraitSimulation, GLM, StatsBase, OrdinalMultinomialModels","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Reading-genotype-data-using-SnpArrays","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Reading genotype data using SnpArrays","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"First use SnpArrays.jl to read in the genotype data. We use PLINK formatted data with the same prefixes for the .bim, .fam, .bed files.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"SnpArrays is a very useful utility and can do a lot more than just read in the data. More information about all the functionality of SnpArrays can be found at: https://openmendel.github.io/SnpArrays.jl/latest/","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"As missing genotypes are often due to problems making the calls, the called genotypes at a marker with too much missing genotypes are potentially unreliable. By default, SnpArrays filters to keep only the genotypes with success rates greater than 0.98 and the minimum minor allele frequency to be 0.01. If the user wishes to change the stringency, change the number given in filter according to SnpArrays.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"filename = \"/mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered\"\nfull_snps = SnpArray(filename * \".bed\");","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"full_snp_data = SnpData(filename)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"snpid = full_snp_data.snp_info[!, :snpid] # store the snp_info with the snp names\ncausal_snp_index = findall(x -> x == \"rs6603811\", snpid)[:] # find the index of the snp of interest by snpid","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"1-element Array{Int64,1}:\n 174","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"The published hypertension GWAS analysis includes the following covariates: sex, center, age, age2, BMI, and the top ten principal components to adjust for ancestry/relatedness.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"causal_snp = @view full_snps[:, causal_snp_index]","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"185565x—1 view(::SnpArray, :, [174]) with eltype UInt8:\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n ...\n 0x03\n 0x03\n 0x03\n 0x03\n 0x02\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"maf(causal_snp)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"1-element Array{Float64,1}:\n 0.05023244010820205","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"locus = convert(Matrix{Float64}, causal_snp, impute = true)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"185565x—1 Array{Float64,2}:\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n ...\n 2.0\n 2.0\n 2.0\n 1.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Power-Calculation","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Power Calculation","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"Now we show how to simulate from customized simulation models that accompany specific genetic analysis options in OpenMendel; for example, ordered, multinomial traits and Variance Component Models.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"This example illustrates the use of the simulations to generates data sets allowing researchers to accurately check the validity of programs and to calculate power for their proposed studies. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"We illustrate this example in three digestable steps: ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"The first by simulating genotypes and covariate values representative for our study population.\nCarry over the simulated design matrix from (1) to create the OrderedMultinomialTrait model object.\nSimulate off the OrderedMultinomialTrait model object created in (2) and run the power analyses for the desired significance level.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"The published hypertension GWAS analysis includes the following covariates: sex, center, age, age2, BMI, and the top ten principal components to adjust for ancestry/relatedness. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"n = length(locus)\npublished_covariate_data = CSV.read(\"/mnt/UKBiobank/ukbdata/ordinalanalysis/Covariate_Final.csv\")\ncovariates = published_covariate_data[:, :]\nsex = Float64.(covariates[!, :sex])","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"185565-element Array{Float64,1}:\n 1.0\n 1.0\n 1.0\n 0.0\n 1.0\n 1.0\n 0.0\n 0.0\n 1.0\n 1.0\n 0.0\n 0.0\n 0.0\n ...  \n 0.0\n 1.0\n 0.0\n 0.0\n 1.0\n 1.0\n 0.0\n 0.0\n 0.0\n 0.0\n 0.0\n 1.0","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"beta_full = \n [0.58\n  0.022]","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"3-element Array{Float64,1}:\n 0.58 \n 0.022\n 0.03","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"X_full = [sex locus]","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"185565Ã—3 Array{Float64,2}:\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n    .      .     .             \n 1.0  0.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  1.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  1.0  2.0","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"link = LogitLink()\ntheta = [1.0, 1.2, 1.4]\n\nOrdinal_Model_Test = OrderedMultinomialTrait(X_full, beta_full, theta, link)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"Ordinal Multinomial Model\n  * number of fixed effects: 3\n  * number of ordinal multinomial outcome categories: 4\n  * link function: LogitLink\n  * sample size: 185565","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"Now, we can do the same procedure for three snps with different minor allele frequencies, and super impose the power plots to compare the trajectory. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"gamma_s = collect(0.0:0.001:0.05)\nplot(gamma_s, m22_power, title = \"Ordered Multinomial Power\", label = \"maf = 0.23\", lw = 3 , legend = :left, legendfontsize= 9)  # plot power\nplot!(gamma_s, m109_power, label = \"SNP: rs76601697, maf = 0.11\", lw = 3 )  # plot power\nplot!(gamma_s, m0_0502_power, label = \"maf = 0.05\", lw = 3 )  # plot power\nxlabel!(\"Odds Ratio for Associated SNP\")\nhline!([.8], label = \"power = 80%, n = 20,000, alpha = 5*10^-8\", lw = 1)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"(Image: png)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/#Citations:","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Citations:","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"[1] Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.`","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"[2] OPENMENDEL: a cooperative programming project for statistical genetics. Hum Genet. 2019 Mar 26. doi: 10.1007/s00439-019-02001-z.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_powerd/","page":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","title":"Trait Simulation - Ordered Multinomial Power (UKBiobank)","text":"[3] German, CA, Sinsheimer, JS, Klimentidis, YC, Zhou, H, Zhou, JJ. Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale. Genetic Epidemiology. 2019; 1â€“ 13. https://doi.org/10.1002/gepi.22276","category":"page"},{"location":"examples/testing_MendelIHT_glm/#Trait-Simulation:-Prototyping-Analysis-Methods-(MendelIHT)","page":"Testing Methods","title":"Trait Simulation: Prototyping Analysis Methods (MendelIHT)","text":"","category":"section"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Authors: Sarah Ji, Benjamin Chu, Janet Sinsheimer, Kenneth Lange, Hua Zhou","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"This software package, TraitSimuliation.jl addresses the need for simulated trait data in genetic analyses.  This package generates data sets that will allow researchers to accurately check the validity of programs and to calculate power for their proposed studies. ","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"In this notebook we show how to the TraitSimulation.jl package can be used to validate the results of the newest OpenMendel analysis package, MendelIHT.jl.","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"First we will generate the desired non-genetic covariates for our examples, sex and age from the GLM.jl package. Then we will make the appropriate calls to OpenMendel packages SnpArrays.jl, MendelIHT.jl. and TraitSimulation.jl to construct the desired genetic model, simulate from it and test the iterative hard thresholding model.","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"For the following three distributions, we simulate traits and validate the results using MendelIHT.","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Normal Trait\nPoisson Trait\nNegative Binomial Trait","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Double check that you are using Julia version 1.0 or higher by checking the machine information","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"versioninfo()","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Julia Version 1.2.0\nCommit c6da87ff4b (2019-08-20 00:03 UTC)\nPlatform Info:\n  OS: macOS (x86_64-apple-darwin18.6.0)\n  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz\n  WORD_SIZE: 64\n  LIBM: libopenlibm\n  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"using Random, DataFrames, LinearAlgebra\nusing SnpArrays, TraitSimulation, GLM, StatsBase, MendelIHT\nusing Revise\nrevise()\n\nRandom.seed!(123);","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Simulate Non-Genetic Covariates","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"n = 5000\np = 10000\ns = 5\n\n# simulate non genetic covariates\nintercept = ones(n, 1) \nsex = rand(Bernoulli(0.51), n)\nage = zscore(rand(Normal(45, 8), n))\nnon_genetic_cov =  [sex age]\nÎ²_non_gen = rand(size(non_genetic_cov, 2))","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"2-element Array{Float64,1}:\n 0.7373240898122129\n 0.4314238505987109","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Simulate Genetic Covariates ","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"First we will generate the entire snparray of p = 10,000 snps for n = 1000 people. Then for s = 10 causal snps, we will convert them to Float64 integers for simulation.","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"# generate entire snparray\nx = simulate_random_snparray(n, p, \"tmp.bed\")\nxbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); \nk = 5\n# generate snparray with only causal snps\ncausal_idx = rand(1:10000, k) # these are the true indices\ncausal_effect_sizes = rand(k) # these are the true effect sizes\n\nx_causal = convert(Matrix{Float64}, @view(x[:, causal_idx]), center = true, scale = true);","category":"page"},{"location":"examples/testing_MendelIHT_glm/#Simulate-the-GLM-Trait","page":"Testing Methods","title":"Simulate the GLM Trait","text":"","category":"section"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Use TraitSimulation.jl to construct the desired model and simulate from it. For the following three distributions, we simulate traits and validate the results using MendelIHT.","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Normal Trait\nPoisson Trait\nNegative Binomial Trait","category":"page"},{"location":"examples/testing_MendelIHT_glm/#Testing-Normal-Trait","page":"Testing Methods","title":"Testing Normal Trait","text":"","category":"section"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"distribution = Normal()\nlink = IdentityLink()\n# revise()\nGLM_model = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, distribution, link)\ny = simulate(GLM_model)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"5000-element Array{Float64,1}:\n  0.9185868876881607    \n -0.9518262679882143    \n -0.33841827510588013   \n  2.049817083313287     \n  1.955913151958258     \n  0.00024397836600864933\n -0.6593742370456934    \n -0.28351854873874294   \n  0.3185448435991667    \n  0.30756420729346534   \n -0.4628928677570747    \n  1.292210040257612     \n  0.06891346416538825   \n  â‹®                     \n -1.0982567760970272    \n -0.016491104287002845  \n  0.9345002196941151    \n  0.14027327315163013   \n -2.20745242227931      \n  1.8222300720529194    \n  2.9752072523810957    \n  0.4504020749944919    \n -1.7701081883560879    \n -0.3661636604161142    \n  2.5102925366869027    \n  0.9083339384177506","category":"page"},{"location":"examples/testing_MendelIHT_glm/#Run-Mendel-IHT","page":"Testing Methods","title":"Run Mendel IHT","text":"","category":"section"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"s_guess = 7\n@show Î²_non_gen\n@show truth = causal_idx \n@show causal_effect_sizes\nresult = L0_reg(x, xbm, non_genetic_cov, y, 50, s_guess, Normal(), IdentityLink(), debias = false, max_iter = 500, verbose= false)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Î²_non_gen = [0.7373240898122129, 0.4314238505987109]\ntruth = causal_idx = [184, 4492, 5382, 161, 2773]\ncausal_effect_sizes = [0.6877212499518375, 0.7610202033321085, 0.6111539989474444, 0.7062198720454749, 0.06323470301651035]\n\n\n\n\n\n\nIHT estimated 5 nonzero SNP predictors and 2 non-genetic predictors.\n\nCompute time (sec):     2.9930989742279053\nFinal loglikelihood:    -7115.458807213961\nIterations:             10\n\nSelected genetic predictors:\n5Ã—2 DataFrame\nâ”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚\nâ”‚     â”‚ \u001b[90mInt64\u001b[39m    â”‚ \u001b[90mFloat64\u001b[39m     â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 161      â”‚ 0.702183    â”‚\nâ”‚ 2   â”‚ 184      â”‚ 0.68639     â”‚\nâ”‚ 3   â”‚ 4492     â”‚ 0.75284     â”‚\nâ”‚ 4   â”‚ 5382     â”‚ 0.595549    â”‚\nâ”‚ 5   â”‚ 5867     â”‚ 0.0581857   â”‚\n\nSelected nongenetic predictors:\n2Ã—2 DataFrame\nâ”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚\nâ”‚     â”‚ \u001b[90mInt64\u001b[39m    â”‚ \u001b[90mFloat64\u001b[39m     â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 1        â”‚ 0.715045    â”‚\nâ”‚ 2   â”‚ 2        â”‚ 0.428766    â”‚","category":"page"},{"location":"examples/testing_MendelIHT_glm/#Testing-Poisson-Trait","page":"Testing Methods","title":"Testing Poisson Trait","text":"","category":"section"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"distribution2 = Poisson()\nlink2 = LogLink()\n\nnon_genetic_cov =  [sex age]\nÎ²_non_gen = [0.1, 0.002]","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"2-element Array{Float64,1}:\n 0.1  \n 0.002","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"GLM_model2 = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, Poisson, link2)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Generalized Linear Model\n  * response distribution: UnionAll\n  * link function: LogLink\n  * sample size: 5000","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"x = simulate_random_snparray(n, p, \"tmp.bed\")\nxbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true);\nk = 5\ncausal_idx = rand(1:10000, k) # these are the true indices\ncausal_effect_sizes = rand(k)\nx_causal = convert(Matrix{Float64}, @view(x[:, causal_idx]), center = true, scale = true);","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"# construct the model \nGLM_model_p2 = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, Poisson(), link2)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Generalized Linear Model\n  * response distribution: Poisson\n  * link function: LogLink\n  * sample size: 5000","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"# simulate the trait\ny_poisson2 = simulate(GLM_model_p2)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"5000-element Array{Int64,1}:\n 0\n 3\n 0\n 0\n 2\n 0\n 0\n 0\n 0\n 7\n 1\n 0\n 2\n â‹®\n 6\n 1\n 2\n 4\n 4\n 2\n 0\n 0\n 1\n 3\n 1\n 0","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"s_guess = 7\nresult2 = L0_reg(x, xbm, non_genetic_cov, Float64.(y_poisson2), 50, s_guess, Poisson(), LogLink(), debias = false, max_iter = 500, verbose= false)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"IHT estimated 6 nonzero SNP predictors and 1 non-genetic predictors.\n\nCompute time (sec):     8.748939990997314\nFinal loglikelihood:    -6742.100029722215\nIterations:             32\n\nSelected genetic predictors:\n6Ã—2 DataFrame\nâ”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚\nâ”‚     â”‚ \u001b[90mInt64\u001b[39m    â”‚ \u001b[90mFloat64\u001b[39m     â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 400      â”‚ 0.405941    â”‚\nâ”‚ 2   â”‚ 2943     â”‚ 0.0423735   â”‚\nâ”‚ 3   â”‚ 3749     â”‚ 0.036553    â”‚\nâ”‚ 4   â”‚ 4656     â”‚ 0.53714     â”‚\nâ”‚ 5   â”‚ 7563     â”‚ 0.722796    â”‚\nâ”‚ 6   â”‚ 8016     â”‚ 0.239444    â”‚\n\nSelected nongenetic predictors:\n1Ã—2 DataFrame\nâ”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚\nâ”‚     â”‚ \u001b[90mInt64\u001b[39m    â”‚ \u001b[90mFloat64\u001b[39m     â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 1        â”‚ 0.0952795   â”‚","category":"page"},{"location":"examples/testing_MendelIHT_glm/#Testing-Negative-Binomial-Trait","page":"Testing Methods","title":"Testing Negative Binomial Trait","text":"","category":"section"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"# simulate the trait\ndistribution = NegativeBinomial()\nlink = LogLink()\n\nGLM_model = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, distribution, link)\ny_negbinomial = simulate(GLM_model)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"5000-element Array{Int64,1}:\n  1\n  1\n  0\n  0\n  2\n  1\n  1\n  1\n  0\n  1\n  0\n  1\n  0\n  â‹\n 15\n  2\n  3\n  2\n 13\n  1\n  0\n  0\n  1\n  1\n  0\n  0","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"# run IHT\ns_guess = 7\n@show Î²_non_gen\n@show truth = causal_idx \n@show causal_effect_sizes\nresult = L0_reg(x, xbm, non_genetic_cov, Float64.(y_negbinomial), 50, s_guess, NegativeBinomial(), link, debias = false, use_maf = false, max_iter = 500, verbose= false)","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"Î²_non_gen = [0.1, 0.002]\ntruth = causal_idx = [4656, 9394, 400, 8016, 7563]\ncausal_effect_sizes = [0.5264504200595785, 0.008085205500707904, 0.4110283000164554, 0.22887974718537518, 0.7299946760225595]\n\n\n\n\n\n\nIHT estimated 6 nonzero SNP predictors and 1 non-genetic predictors.\n\nCompute time (sec):     4.818864107131958\nFinal loglikelihood:    -7579.647514751706\nIterations:             17\n\nSelected genetic predictors:\n6Ã—2 DataFrame\nâ”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚\nâ”‚     â”‚ \u001b[90mInt64\u001b[39m    â”‚ \u001b[90mFloat64\u001b[39m     â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 400      â”‚ 0.418851    â”‚\nâ”‚ 2   â”‚ 1118     â”‚ 0.0727529   â”‚\nâ”‚ 3   â”‚ 3367     â”‚ -0.0822334  â”‚\nâ”‚ 4   â”‚ 4656     â”‚ 0.50584     â”‚\nâ”‚ 5   â”‚ 7563     â”‚ 0.737677    â”‚\nâ”‚ 6   â”‚ 8016     â”‚ 0.193797    â”‚\n\nSelected nongenetic predictors:\n1Ã—2 DataFrame\nâ”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚\nâ”‚     â”‚ \u001b[90mInt64\u001b[39m    â”‚ \u001b[90mFloat64\u001b[39m     â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 1        â”‚ 0.140459    â”‚","category":"page"},{"location":"examples/testing_MendelIHT_glm/#Citations:","page":"Testing Methods","title":"Citations:","text":"","category":"section"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"[1] Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.`","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"[2] OPENMENDEL: a cooperative programming project for statistical genetics. Hum Genet. 2019 Mar 26. doi: 10.1007/s00439-019-02001-z.","category":"page"},{"location":"examples/testing_MendelIHT_glm/","page":"Testing Methods","title":"Testing Methods","text":"[3] Benjamin B. Chu, Kevin L. Keys, Christopher A. German, Hua Zhou, Jin J. Zhou, Janet S. Sinsheimer, Kenneth Lange. Iterative Hard Thresholding in GWAS: Generalized Linear Models, Prior Weights, and Double Sparsity. bioRxiv doi:10.1101/697755","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Trait-Simulation-Ordinal-Multinomial-Power","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Authors: Sarah Ji, Janet Sinsheimer, Kenneth Lange, Hua Zhou","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"In this notebook we show how to use the TraitSimulation.jl package we illustrate how TraitSimulation.jl can easily simulate traits from genotype data, all within the OpenMendel universe. Operating within this universe brings potential advantages over the available software(s) when needed for downstream analysis or study design. Using just a few calls on the command line to the appropriate packages within the OpenMendel, we demonstrate how to use the OpenMendel code pipeline to construct the genetic model, simulate the trait and find the power for the Ordinal Multinomial Trait. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Background","page":"Trait Simulation - Ordinal Multinomial Power","title":"Background","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"The ordinal multinomial model is a powerful way to model stages of disease progression. For a disease like diabetes, the monitoring of the disease's progression may be improved based on several different biomarkers detected in the EHR data, such as diabetes diagnostic codes, diabetes medication, hyperglycemia in blood results defined by HbA1c and fasting glucose levels, and presence of diabetes process of care codes. Then these algorithm can better categorize individuals into disease progression categories that relate to how likely they are to have diabetes. This is motivated by the need for early diagnoses and effective intervention techniques. For a detailed review on the published ordinal model, we provide in the references our most recent published paper in Genetic Epidemiology, Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"This software package, TraitSimuliation.jl addresses the need for simulated trait data in genetic analyses.  This package generates data sets that will allow researchers to accurately check the validity of programs and to calculate power for their proposed studies. This package gives users the ability to easily simulate phenotypic traits under generalized linear models (GLMs) or variance component models (VCMs) conditional on PLINK formatted genotype data [3]. In addition, we demo customized simulation utilities that accompany specific genetic analysis options in Open-Mendel; for example, ordered, multinomial traits. We demonstrate these simulation utilities on the example dataset described below.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Demonstration","page":"Trait Simulation - Ordinal Multinomial Power","title":"Demonstration","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Example-Data","page":"Trait Simulation - Ordinal Multinomial Power","title":"Example Data","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"The data in this demo is restricted to a subset of individuals of European ancestry excluding first- and second-degree relatives based off of kinship estimates. We filtered samples by 98% genotyping success rate on all chromosomes and SNPs by 99% genotyping success rate. These measures result in n = 185,565 individuals and 470,228 SNPs for analysis. A more detailed description of the data can be found in the references 3 The data from this notebook can be found in the UK Biobank repository under Project ID: 48152 and 15678 at https://www.ukbiobank.ac.uk with the permission of UK Biobank. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"We use the OpenMendel package SnpArrays.jl to both read in and write out PLINK formatted files. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Double-check-that-you-are-using-Julia-version-1.0-or-higher-by-checking-the-machine-information","page":"Trait Simulation - Ordinal Multinomial Power","title":"Double check that you are using Julia version 1.0 or higher by checking the machine information","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"versioninfo()","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Julia Version 1.3.0\nCommit 46ce4d7933 (2019-11-26 06:09 UTC)\nPlatform Info:\n  OS: Linux (x86_64-pc-linux-gnu)\n  CPU: Intel(R) Core(TM) i9-9920X CPU @ 3.50GHz\n  WORD_SIZE: 64\n  LIBM: libopenlibm\n  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"using Plots, DataFrames, LinearAlgebra, StatsFuns, CSV\nusing SnpArrays, TraitSimulation, GLM, StatsBase, OrdinalMultinomialModels","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Reading-genotype-data-using-SnpArrays","page":"Trait Simulation - Ordinal Multinomial Power","title":"Reading genotype data using SnpArrays","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"First use SnpArrays.jl to read in the genotype data. We use PLINK formatted data with the same prefixes for the .bim, .fam, .bed files.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"SnpArrays is a very useful utility and can do a lot more than just read in the data. More information about all the functionality of SnpArrays can be found at: https://openmendel.github.io/SnpArrays.jl/latest/","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"As missing genotypes are often due to problems making the calls, the called genotypes at a marker with too much missing genotypes are potentially unreliable. By default, SnpArrays filters to keep only the genotypes with success rates greater than 0.98 and the minimum minor allele frequency to be 0.01. If the user wishes to change the stringency, change the number given in filter according to SnpArrays.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"filename = \"/mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered\"\nfull_snps = SnpArray(filename * \".bed\");","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"full_snp_data = SnpData(filename)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"SnpData(people: 185565, snps: 470228,\nsnp_info: \nâ”‚ Row â”‚ chromosome â”‚ snpid       â”‚ genetic_distance â”‚ position â”‚ allele1      â”‚ allele2      â”‚\nâ”‚     â”‚ String     â”‚ String      â”‚ Float64          â”‚ Int64    â”‚ Categoricalâ€¦ â”‚ Categoricalâ€¦ â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 1          â”‚ rs3131972   â”‚ 0.0              â”‚ 752721   â”‚ A            â”‚ G            â”‚\nâ”‚ 2   â”‚ 1          â”‚ rs12184325  â”‚ 0.0              â”‚ 754105   â”‚ T            â”‚ C            â”‚\nâ”‚ 3   â”‚ 1          â”‚ rs3131962   â”‚ 0.0              â”‚ 756604   â”‚ A            â”‚ G            â”‚\nâ”‚ 4   â”‚ 1          â”‚ rs12562034  â”‚ 0.0              â”‚ 768448   â”‚ A            â”‚ G            â”‚\nâ”‚ 5   â”‚ 1          â”‚ rs116390263 â”‚ 0.0              â”‚ 772927   â”‚ T            â”‚ C            â”‚\nâ”‚ 6   â”‚ 1          â”‚ rs4040617   â”‚ 0.0              â”‚ 779322   â”‚ G            â”‚ A            â”‚\nâ€¦,\nperson_info: \nâ”‚ Row â”‚ fid       â”‚ iid       â”‚ father    â”‚ mother    â”‚ sex       â”‚ phenotype â”‚\nâ”‚     â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 1000019   â”‚ 1000019   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 2   â”‚ 1000078   â”‚ 1000078   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 3   â”‚ 1000081   â”‚ 1000081   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 4   â”‚ 1000105   â”‚ 1000105   â”‚ 0         â”‚ 0         â”‚ 2         â”‚ -9        â”‚\nâ”‚ 5   â”‚ 1000112   â”‚ 1000112   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 6   â”‚ 1000129   â”‚ 1000129   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ€¦,\nsrcbed: /mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered.bed\nsrcbim: /mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered.bim\nsrcfam: /mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered.fam\n)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"snpid = full_snp_data.snp_info[!, :snpid] # store the snp_info with the snp names\ncausal_snp_index = findall(x -> x == \"rs6603811\", snpid)[:] # find the index of the snp of interest by snpid","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"1-element Array{Int64,1}:\n 174","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"The published hypertension GWAS analysis includes the following covariates: sex, center, age, age2, BMI, and the top ten principal components to adjust for ancestry/relatedness.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"causal_snp = @view full_snps[:, causal_snp_index]","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"185565Ã—1 view(::SnpArray, :, [174]) with eltype UInt8:\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n    â‹®\n 0x03\n 0x03\n 0x03\n 0x02\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03\n 0x03","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"maf(causal_snp)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"1-element Array{Float64,1}:\n 0.05023244010820205","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"locus = convert(Matrix{Float64}, causal_snp, impute = true)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"185565Ã—1 Array{Float64,2}:\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n â‹®  \n 2.0\n 2.0\n 2.0\n 1.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0\n 2.0","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Power-Calculation","page":"Trait Simulation - Ordinal Multinomial Power","title":"Power Calculation","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Now we show how to simulate from customized simulation models that accompany specific genetic analysis options in OpenMendel; for example, ordered, multinomial traits and Variance Component Models.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"This example illustrates the use of the simulations to generates data sets allowing researchers to accurately check the validity of programs and to calculate power for their proposed studies. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"We illustrate this example in three digestable steps: ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"The first by simulating genotypes and covariate values representative for our study population.\nCarry over the simulated design matrix from (1) to create the OrderedMultinomialTrait model object.\nSimulate off the OrderedMultinomialTrait model object created in (2) and run the power analyses for the desired significance level.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"The published hypertension GWAS analysis includes the following covariates: sex, center, age, age2, BMI, and the top ten principal components to adjust for ancestry/relatedness. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"n = length(locus)\npublished_covariate_data = CSV.read(\"/mnt/UKBiobank/ukbdata/ordinalanalysis/Covariate_Final.csv\")\ncovariates = published_covariate_data[:, :]\nsex = Float64.(covariates[!, :sex])","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"185565-element Array{Float64,1}:\n 1.0\n 1.0\n 1.0\n 0.0\n 1.0\n 1.0\n 0.0\n 0.0\n 1.0\n 1.0\n 0.0\n 0.0\n 0.0\n â‹®  \n 0.0\n 1.0\n 0.0\n 0.0\n 1.0\n 1.0\n 0.0\n 0.0\n 0.0\n 0.0\n 0.0\n 1.0","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Î²_full = \n [0.58\n  0.022]","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"3-element Array{Float64,1}:\n 0.58 \n 0.022\n 0.03","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Î³s = collect(0.02:0.005:0.055)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"8-element Array{Float64,1}:\n 0.02 \n 0.025\n 0.03 \n 0.035\n 0.04 \n 0.045\n 0.05 \n 0.055","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"X_full = [sex locus]","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"185565Ã—3 Array{Float64,2}:\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n â‹®            \n 1.0  0.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  1.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  0.0  2.0\n 1.0  1.0  2.0","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"link = LogitLink()\nÎ¸ = [1.0, 1.2, 1.4]\n\nOrdinal_Model_Test = OrderedMultinomialTrait(X_full, Î²_full, Î¸, link)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Ordinal Multinomial Model\n  * number of fixed effects: 3\n  * number of ordinal multinomial outcome categories: 4\n  * link function: LogitLink\n  * sample size: 185565","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Now, we can do the same procedure for three snps with different minor allele frequencies, and super impose the power plots to compare the trajectory. ","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"Î³s = collect(0.0:0.001:0.05)\nplot(Î³s, m22_power, title = \"Ordered Multinomial Power\", label = \"maf = 0.23\", lw = 3 , legend = :left, legendfontsize= 9)  # plot power\nplot!(Î³s, m109_power, label = \"SNP: rs76601697, maf = 0.11\", lw = 3 )  # plot power\nplot!(Î³s, m0_0502_power, label = \"maf = 0.05\", lw = 3 )  # plot power\nxlabel!(\"Odds Ratio for Associated SNP\")\nhline!([.8], label = \"power = 80%, n = 20,000, alpha = 5*10^-8\", lw = 1)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"(Image: png)","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/#Citations:","page":"Trait Simulation - Ordinal Multinomial Power","title":"Citations:","text":"","category":"section"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"[1] Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.`","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"[2] OPENMENDEL: a cooperative programming project for statistical genetics. Hum Genet. 2019 Mar 26. doi: 10.1007/s00439-019-02001-z.","category":"page"},{"location":"examples/ukbiobank_ordered_multinomial_power/","page":"Trait Simulation - Ordinal Multinomial Power","title":"Trait Simulation - Ordinal Multinomial Power","text":"[3] German, CA, Sinsheimer, JS, Klimentidis, YC, Zhou, H, Zhou, JJ. Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale. Genetic Epidemiology. 2019; 1â€“ 13. https://doi.org/10.1002/gepi.22276","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Trait-Simulation-Variance-Component-Model-Power-(UKBiobank)","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Authors: Sarah Ji, Janet Sinsheimer, Kenneth Lange, Hua Zhou, Eric Sobel","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"In this notebook we show how to use the TraitSimulation.jl package we illustrate how TraitSimulation.jl can easily simulate traits from genotype data, all within the OpenMendel universe. Operating within this universe brings potential advantages over the available software(s) when needed for downstream analysis or study design. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Using just a few calls on the command line to the appropriate packages within the OpenMendel, we demonstrate in three easy examples the utilities of the TraitSimulation.jl package.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Users who want a reference on genetic modeling, we recommend  Mathematical And Statistical Methods For Genetic Analysis by Dr. Kenneth Lange. In chapter 8 of this book, the user can find an introduction to Variance Component Models in Genetic Setting. For a more in depth review of variance component modeling in the genetic setting, we include a reference at the end of the notebook [4].","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Background","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Background","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"There is a lack of software available to geneticists who wish to calculate power and sample sizes in designing a study on genetics data. Typically, the study power depends on assumptions about the underlying disease model.  Many power calculating software tools operate as a black box and do not allow for customization.  To develop custom tests, researchers can develop their own simulation procedures to carry out power calculations.  One limitation with many existing methods for simulating traits conditional on genotypes is that these methods are limited to normally distributed traits and to fixed effects. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"This software package, TraitSimuliation.jl addresses the need for simulated trait data in genetic analyses.  This package generates data sets that will allow researchers to accurately check the validity of programs and to calculate power for their proposed studies. We demonstrate customized simulation utilities that accompany specific genetic analysis options in Open-Mendel; for example Variance Component Models (VCM) and ordered multinomial traits. We demonstrate the simulation utilities on a subset of 20,000 individuals and 470228 snps in the UK Biobank dataset. We use a subset of the dataset in the Ordinal GWAS paper to include only the first 20,000 individuals for this demo. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Demonstration","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Demonstration","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/#Example-Data","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Example Data","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"We use the OpenMendel package SnpArrays.jl to both read in and write out PLINK formatted files. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"For convenience we use the common assumption that the residual covariance among two relatives can be captured by the additive genetic variance times twice the kinship coefficient.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"In each example the user can specify the simulation model parameters, along with the number of repetitions for each simulation model as desired. By default, the simulation will return the result of a single simulation.","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Double-check-that-you-are-using-Julia-version-1.0-or-higher-by-checking-the-machine-information","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Double check that you are using Julia version 1.0 or higher by checking the machine information","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"versioninfo()","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Julia Version 1.3.0\nCommit 46ce4d7933 (2019-11-26 06:09 UTC)\nPlatform Info:\n  OS: Linux (x86_64-pc-linux-gnu)\n  CPU: Intel(R) Core(TM) i9-9920X CPU @ 3.50GHz\n  WORD_SIZE: 64\n  LIBM: libopenlibm\n  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"using SnpArrays, Distributions, Test, Statistics, VarianceComponentModels, Distributions, Random\nusing DataFrames, LinearAlgebra, StatsBase, TraitSimulation, Distributions, CSV, Plots, BenchmarkTools\nRandom.seed!(1234);","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Reading-genotype-data-using-SnpArrays","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Reading genotype data using SnpArrays","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"First use SnpArrays.jl to read in the genotype data. We use PLINK formatted data with the same prefixes for the .bim, .fam, .bed files.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"SnpArrays is a very useful utility and can do a lot more than just read in the data. More information about all the functionality of SnpArrays can be found at: https://openmendel.github.io/SnpArrays.jl/latest/","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"As missing genotypes are often due to problems making the calls, the called genotypes at a marker with too much missing genotypes are potentially unreliable. By default, SnpArrays filters to keep only the genotypes with success rates greater than 0.98 and the minimum minor allele frequency to be 0.01. If the user wishes to change the stringency, change the number given in filter according to SnpArrays.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"filename = \"/mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered\"\nfull_snp_data = SnpData(filename)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"SnpData(people: 185565, snps: 470228,\nsnp_info: \nâ”‚ Row â”‚ chromosome â”‚ snpid       â”‚ genetic_distance â”‚ position â”‚ allele1      â”‚ allele2      â”‚\nâ”‚     â”‚ String     â”‚ String      â”‚ Float64          â”‚ Int64    â”‚ Categoricalâ€¦ â”‚ Categoricalâ€¦ â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 1          â”‚ rs3131972   â”‚ 0.0              â”‚ 752721   â”‚ A            â”‚ G            â”‚\nâ”‚ 2   â”‚ 1          â”‚ rs12184325  â”‚ 0.0              â”‚ 754105   â”‚ T            â”‚ C            â”‚\nâ”‚ 3   â”‚ 1          â”‚ rs3131962   â”‚ 0.0              â”‚ 756604   â”‚ A            â”‚ G            â”‚\nâ”‚ 4   â”‚ 1          â”‚ rs12562034  â”‚ 0.0              â”‚ 768448   â”‚ A            â”‚ G            â”‚\nâ”‚ 5   â”‚ 1          â”‚ rs116390263 â”‚ 0.0              â”‚ 772927   â”‚ T            â”‚ C            â”‚\nâ”‚ 6   â”‚ 1          â”‚ rs4040617   â”‚ 0.0              â”‚ 779322   â”‚ G            â”‚ A            â”‚\nâ€¦,\nperson_info: \nâ”‚ Row â”‚ fid       â”‚ iid       â”‚ father    â”‚ mother    â”‚ sex       â”‚ phenotype â”‚\nâ”‚     â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚ Abstractâ€¦ â”‚\nâ”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤\nâ”‚ 1   â”‚ 1000019   â”‚ 1000019   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 2   â”‚ 1000078   â”‚ 1000078   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 3   â”‚ 1000081   â”‚ 1000081   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 4   â”‚ 1000105   â”‚ 1000105   â”‚ 0         â”‚ 0         â”‚ 2         â”‚ -9        â”‚\nâ”‚ 5   â”‚ 1000112   â”‚ 1000112   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ”‚ 6   â”‚ 1000129   â”‚ 1000129   â”‚ 0         â”‚ 0         â”‚ 1         â”‚ -9        â”‚\nâ€¦,\nsrcbed: /mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered.bed\nsrcbim: /mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered.bim\nsrcfam: /mnt/UKBiobank/ukbdata/ordinalanalysis/ukb.plink.filtered.fam\n)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"sample_snps = SnpArray(\"tmp.filter.20k.chr1to7.bed\", 20000)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"20000X—425292 SnpArray:\n 0x02  0x02  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x02  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x02  0x02  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x00  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x02  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x02  0x03  0x02  0x03  0x03  0x02    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x02  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x02  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x02  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x02  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x02  0x03  0x02  0x03  0x03  0x02    0x00  0x00  0x00  0x00  0x00  0x00\n 0x02  0x03  0x02  0x03  0x03  0x02    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x02  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x03  0x03  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x02  0x02  0x03  0x03  0x03  0x03    0x00  0x00  0x00  0x00  0x00  0x00\n 0x02  0x03  0x02  0x03  0x03  0x02    0x00  0x00  0x00  0x00  0x00  0x00","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"The published hypertension GWAS analysis includes the following covariates: sex, center, age, age2, BMI, and the top ten principal components to adjust for ancestry/relatedness.","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Construct-Genetic-Model:-(VCM-Parameter-Specification)","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Construct Genetic Model: (VCM Parameter Specification)","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Say that you have the the classical setting in genetics, two variance components, one for the additive genetic variance and one for the environmental variance. For simulating d correlated traits for n related people under the VCM, users can specify their covariance structures as follows.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Y_n times d sim textMatrixNormal(mathbfM_n times d = XB Omega_nd times nd = Sigma_A otimes V_A + Sigma_E otimes V_E)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Suppose we want to see a particular SNP has an effect on a given phenotype after accounting for relatedness among individuals. Here we fit variance component model with a single SNP 's' as fixed effect. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"hspace5em  mathbfy = mathbfXmathbfbeta + mathbfG_s gamma + mathbfg + mathbfepsilon hspace5em (1)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"$","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"\\mathbf{g} \\sim N(\\mathbf{0}, \\sigmag^2\\mathbf{\\Phi}); \\mathbf{\\epsilon} \\sim N(\\mathbf{0}, \\sigmae^2\\mathbf{I}) (2) $","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"where  mathbfy: phenotype and ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Fixed effects:    mathbfX textmatrix of covariates including intercept   beta textvector of covariate effects including intercept   mathbfG_s textgenotype of SNP s   gamma text(scalar) association parameter of interest measuring the effect of genotype on phenotype\nRandom effects:   mathbfg textrandom vector of polygenic effects with mathbfg sim N(mathbf0 sigma_g^2 mathbfPhi)       sigma_g^2 textadditive genetic variance       mathbfPhi textmatrix of pairwise measures of genetic relatedness   epsilon textrandom vector with epsilon sim N(mathbf0 sigma_e^2mathbfI)       sigma_e^2 textnon-genetic variance due to non-genetic effects assumed to be acting independently on individuals","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"To test whether SNP 's' is associated with phenotype, we fit two models. First consider the model without SNP  's' as fixed effects (aka null model): ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"hspace5em  mathbfy = mathbfXmathbfbeta + mathbfg + mathbfepsilon hspace5em (2)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"and the model with SNP 's' as fixed effects (1). Then we can compare the log likelihood to see if there is improvement in the model fit with inclusion of the SNP of interest. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"sample_snps = SnpArray(\"tmp.filter.20k.chr1to7.bed\", 20000)\nbimfile = full_snp_data.snp_info # store the snp_info with the snp names\nsnpid  = bimfile[:, :snpid] # store the snp names in the snpid vector\ncausal_snp_index = findfirst(x -> x == \"rs11240779\", snpid); # find the index of the snp of interest by snpid","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"GRM = grm(sample_snps, cinds = (1:50000), minmaf = 0.05)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"20000Ã—20000 Array{Float64,2}:\n  0.484156      0.00597776   -0.00546745   â€¦  -0.000619711  -0.00303873 \n  0.00597776    0.495968     -0.00707199       0.0117363     0.000888748\n -0.00546745   -0.00707199    0.496102        -0.000331281   0.000782371\n  0.00381941    0.00167924    0.00956107       0.000762047   0.00437559 \n -0.00144129   -0.00908248   -0.00472724      -0.00618115   -0.00571785 \n  0.00073305    0.00622141    0.0027326    â€¦  -0.00318683   -0.00419594 \n  0.00314762   -0.00238743    0.000807163      0.00242279   -0.00666091 \n -0.00580214    0.00263065    0.00160677      -0.00371358    0.00494184 \n -0.00075661    0.00225328   -0.00374283      -0.00402865   -0.00304101 \n  0.00229812    0.0054374     0.000593526     -0.00236298   -0.00359522 \n  0.00316909    0.000840955   0.0127431    â€¦  -0.00213737    0.00333481 \n  0.00146024   -0.00469578    0.00319257       0.00558896   -0.000461851\n -0.00324931    0.00241239    0.00797953      -0.0102366    -0.00657798 \n  â‹®                                        â‹±                            \n  0.00164156   -0.00319575   -0.00399023       0.00439952   -0.00151732 \n -0.00761124   -0.00439661    0.000206945      0.00108872   -0.00603369 \n  0.000619589   0.00316643   -0.00132394   â€¦   0.00661205    0.00313253 \n -0.000544604  -0.00491485    0.000410359     -0.0021603     0.00195482 \n  0.000969757  -0.00319238    0.00395337       0.00332249    1.72426e-5 \n -0.00218307   -0.00343704   -0.0019731       -0.00263199   -0.000637567\n  0.00507449   -0.000652566  -0.00806479       0.00143266   -0.00657432 \n -0.0029458     0.00347321   -0.00135441   â€¦   0.00335228    0.00388207 \n -0.0103544    -0.0043632     0.00665735      -0.00523865   -0.00983538 \n  0.00177748    0.00161656   -0.00428977       0.00767306   -0.0046504  \n -0.000619711   0.0117363    -0.000331281      0.495208     -4.80061e-6 \n -0.00303873    0.000888748   0.000782371     -4.80061e-6    0.503932","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Power-Calculation","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Power Calculation","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"This example illustrates the use of the simulations to generates data sets allowing researchers to accurately check the validity of programs and to calculate power for their proposed studies. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"We illustrate this example in three digestable steps: ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"The first by simulating genotypes and covariate values representative for our study population.\nCarry over the simulated design matrix from (1) to create the VCMTrait object.\nSimulate off the VCMTrait model object created in (2) and run the power analyses for the desired significance level.","category":"page"},{"location":"examples/ukbiobank_vcm_power/#convert","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"convert","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"By default convert function translates genotypes according to the additive SNP model, which essentially counts the number of A2 allele (0, 1 or 2) per genotype. Other SNP models are dominant and recessive, both in terms of the A2 allele.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Genotype SnpArray model=ADDITIVE_MODEL model=DOMINANT_MODEL model=RECESSIVE_MODEL\nA1,A1 0x00 0 0 0\nmissing 0x01 NaN NaN NaN\nA1,A2 0x02 1 1 0\nA2,A2 0x03 2 1 1","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"If desired, the user can decide to specify alternative model parameters found in the SnpArrays documentation.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"locus = convert(Vector{Float64}, @view sample_snps[:, causal_snp_index]; impute =  true)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"20000-element Array{Float64,1}:\n 2.0\n 2.0\n 1.0\n 1.0\n 2.0\n 2.0\n 0.0\n 2.0\n 2.0\n 2.0\n 0.0\n 1.0\n 2.0\n â‹® \n 2.0\n 2.0\n 1.0\n 2.0\n 2.0\n 1.0\n 1.0\n 2.0\n 1.0\n 2.0\n 2.0\n 1.0","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Design-Matrix","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Design Matrix","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Now we concatenate the simulated causal locus, transformed to Float64, to the other covariates we want to control for in the study design.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"The published hypertension GWAS analysis includes the following covariates: sex, center, age, age2, BMI, and the top ten principal components to adjust for ancestry/relatedness.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"n = length(locus)\npublished_covariate_data = CSV.read(\"/mnt/UKBiobank/ukbdata/ordinalanalysis/Covariate_Final.csv\")\ncovariates = published_covariate_data[1:20000, :]\nsex = Float64.(covariates[!, :sex])","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Phenotype-Simulation:","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Phenotype Simulation:","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Now that we have our simulated design matrix with the desired SNP of interest, we can simulate our phenotypes on the simulated data under different TraitSimulation models. To illustrate, we demonstrate on the VCMTrait model object in TraitSimulation.jl.","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Univariate-VCM-Trait","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Univariate VCM Trait","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"X = [sex locus]\nÎ²_full =\n [0.52\n0.05]\n\nI_n = Matrix(I, size(GRM))\nvc = @vc [0.11][:, :]  âŠ— 2GRM + [1.0][:, :] âŠ— I_n;\ntrait = VCMTrait(X, Î²_full[:, :], vc)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Variance Component Model\n  * number of traits: 1\n  * number of variance components: 2\n  * sample size: 20000","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Bivariate-VCM-Trait","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Bivariate VCM Trait","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Î£A = [0.3 0.0; 0.0 0.3]\nÎ£E = [1.0 0.0; 0.0 1.0]\nÎ²_full =\n [0.52 0.52\n0.05 0.01]","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"2Ã—2 Array{Float64,2}:\n 0.52  0.52\n 0.05  0.01","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"X = [sex locus]\nI_n = Matrix(I, size(GRM))\nvc = @vc Î£A âŠ— 2GRM + Î£E âŠ— I_n;","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"trait = VCMTrait(X, Î²_full[:, :], vc)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Variance Component Model\n  * number of traits: 2\n  * number of variance components: 2\n  * sample size: 20000","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"We use the following function to generate the p-values for the simulated power example for the ordered multinomial regression model. We range effect sizes in the vector gamma_s, which collects effect sizes from 0 to 0.054. To demo, we simulate the bivariate trait symmetrically using the same single variant and effect size. As expected, the power increases as the effect size increases. ","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"The power calculation works in the following two steps:","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"For each of the specified number of simulations, nsim, we will simulate the bivariate trait under the specified genetic model and effect size. \nFor each simulated phenotype, we use VarianceComponentModels.jl to test the significance of the causal locus.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Y = zeros(size(trait.Î))\nX = trait.X\nV1 = trait.vc[1].V\nV2 = trait.vc[2].V;","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"# allocate variance component objects for null and alternative hypothesis\nvcm_null, vcrot_null, vcm_alt, vcrot_alt = @time null_and_alternative_vcm_and_rotate(Y, X, V1, V2);","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"gamma_s = [collect(0.0:0.005:0.015); collect(0.016:0.002:0.032); collect(0.034:0.005:0.065)]","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"20-element Array{Float64,1}:\n 0.0  \n 0.005\n 0.01 \n 0.015\n 0.016\n 0.018\n 0.02 \n 0.022\n 0.024\n 0.026\n 0.028\n 0.03 \n 0.032\n 0.034\n 0.039\n 0.044\n 0.049\n 0.054\n 0.059\n 0.064","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Univariate-VCM-Power","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Univariate VCM Power","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"For 20,000 individuals at 20 SNP effect sizes each performing 1000 Likelihood Ratio tests while controlling for the estimate for the kinship matrix, it took 4.83 hours to run the following bivariate trait power analysis","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Random.seed!(12345)\nnsim = 1000\npvalues = @time power_simulation(trait, gamma_s, nsim, vcm_null, vcrot_null, vcm_alt, vcrot_alt)\nresults = DataFrame(pvalues)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"******************************************************************************\nThis program contains Ipopt, a library for large-scale nonlinear optimization.\n Ipopt is released as open source code under the Eclipse Public License (EPL).\n         For more information visit http://projects.coin-or.org/Ipopt\n******************************************************************************\n\n17414.656838 seconds (1.93 G allocations: 130.153 GiB, 0.14% gc time)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Each column of this matrix represents each of the detected effect sizes, and each row of this matrix represents each simulation for that effect size. The user feeds into the function the number of simulations, the vector of effect sizes, the TraitSimulation.jl model object, and the random seed.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"For VCMTrait objects, we include additional power utilities that make the appropriate function calls to the VarianceComponentModels module to get the p-value obtained from testing the significance of the causal locus on the simulated trait.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"For each effect size in gamma_s in each column we have the p-values obtained from testing the significance of the causal locus on both traits, nsim = 1000 times under the VCM and the randomseed = 1234.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Now we find the power of each effect size in the user-specified gamma_s vector at the specified alpha level of significance, and plot the trajectory using the Plots.jl package.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"alpha = 5*10^-8\n\nunivariate_power_effectsize = power(pvalues, alpha)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"20-element Array{Float64,1}:\n 0.0  \n 0.0  \n 0.0  \n 0.005\n 0.004\n 0.011\n 0.027\n 0.06 \n 0.103\n 0.187\n 0.288\n 0.423\n 0.592\n 0.688\n 0.913\n 0.983\n 0.998\n 1.0 \n 1.0\n 1.0","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"maf = 0.23\ntotal_variance = 1.3\nprop_v2 = gamma_s.^2 ./ total_variance\nprop_explained_by_SNP = (2*(1 - maf)*maf).* prop_v2\nplot([prop_explained_by_SNP], [univariate_power_effectsize], title = \"Univariate Variance Component Model Power\", label = \"maf = 0.23, snp: rs11240779\", lw = 3 , legend = :bottomright, legendfontsize= 9)  # plot power\nxlabel!(\"Proportion of Variation Explained by the SNP for Each Trait\")\nhline!([.8], label = \"power = 80%, p-value threshold = 5*10^-8\", lw = 3)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"(Image: png)","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Bivariate-VCM-Power","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Bivariate VCM Power","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"For 20,000 individuals at 20 SNP effect sizes each performing 1000 Likelihood Ratio tests while controlling for the estimate for the kinship matrix, it took 5.97 hours to run the following bivariate trait power analysis","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"Random.seed!(12345)\nnsim = 1000\npvalues = @time power_simulation(trait, gamma_s, nsim, vcm_null, vcrot_null, vcm_alt, vcrot_alt)\nresults = DataFrame(pvalues)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"******************************************************************************\nThis program contains Ipopt, a library for large-scale nonlinear optimization.\n Ipopt is released as open source code under the Eclipse Public License (EPL).\n         For more information visit http://projects.coin-or.org/Ipopt\n******************************************************************************\n\n21500.556177 seconds (1.80 G allocations: 164.017 GiB, 0.15% gc time)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"alpha = 5*10^-8\n\nbivariate_power_effectsize = power(pvalues, alpha)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"20-element Array{Float64,1}:\n 0.0  \n 0.0  \n 0.0  \n 0.022\n 0.028\n 0.061\n 0.164\n 0.29 \n 0.473\n 0.652\n 0.83 \n 0.926\n 0.964\n 0.994\n 1.0  \n 1.0  \n 1.0  \n 1.0  \n 1.0\n 1.0","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"maf = 0.23\ntotal_variance = 1.3\nprop_v2 = gamma_s.^2 ./ total_variance\nprop_explained_by_SNP = (2*(1 - maf)*maf).* prop_v2\nplot([prop_explained_by_SNP], [bivariate_power_effectsize], title = \"Bivariate Variance Component Model Power\", label = \"maf = 0.23, snp: rs11240779\", lw = 3 , legend = :bottomright, legendfontsize= 9)  # plot power\nxlabel!(\"Proportion of Variation Explained by the SNP for Each Trait\")\nhline!([.8], label = \"power = 80%, p-value threshold = 5*10^-8\", lw = 3)","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"(Image: png)","category":"page"},{"location":"examples/ukbiobank_vcm_power/#Citations:","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Citations:","text":"","category":"section"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"[1] Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.`","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"[2] OPENMENDEL: a cooperative programming project for statistical genetics. Hum Genet. 2019 Mar 26. doi: 10.1007/s00439-019-02001-z.","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"[3] German, CA, Sinsheimer, JS, Klimentidis, YC, Zhou, H, Zhou, JJ. Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale. Genetic Epidemiology. 2019; 1€“ 13. https://doi.org/10.1002/gepi.22276","category":"page"},{"location":"examples/ukbiobank_vcm_power/","page":"Trait Simulation - Variance Component Model Power (UKBiobank)","title":"Trait Simulation - Variance Component Model Power (UKBiobank)","text":"[4] Lange K, Boehnke M (1983) Extensions to pedigree analysis. IV. Covariance component models for multivariate traits. Amer J Med Genet 14:513€“524","category":"page"},{"location":"#TraitSimulation.jl","page":"Home","title":"TraitSimulation.jl","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Authors: Sarah Ji, Kenneth Lange, Janet Sinsheimer, Jin Zhou, Hua Zhou, Eric Sobel","category":"page"},{"location":"","page":"Home","title":"Home","text":"We present TraitSimulation, an open-source Julia package that makes it trivial to quickly simulate phenotypes under a variety of genetic architectures. This package is integrated into our larger OpenMendel suite of software tools for easy downstream analyses. Julia was purpose-built for scientific programming and provides tremendous speed and memory efficiency, extensive libraries, and easy access to multi-CPU and GPU hardware and to distributed cloud-based parallelization. It is designed to encourage flexible trait simulation with the standard devices of applied statistics, generalized linear models (GLMs)and generalized linear mixed models (GLMMs).TraitSimulation also accommodates many study designs: unrelateds, sibships, pedigrees, or a mixture of all three. ","category":"page"},{"location":"#Background","page":"Home","title":"Background","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Statistical geneticists employ simulation to estimate the power of proposed studies, test new analysis tools, and evaluate properties of causal models. Although there are existing trait simulators, there is ample room for modernization of currently available simulation models and computing platforms.For example, most phenotype simulators are limited to Gaussian traits or traits transformable to normality, ignoring qualitative traits and realistic, non-normal trait distributions. Modern computer languages, such as Julia, that accommodate parallelization and cloud-based computing are now mainstream but rarely used inolder applications. To meet the challenges of contemporary big studies, it is nearly imperative for geneticists to adopt new computational tools. Although GPU and Threading capabilities are relatively new and under current development, we explore the potential of these language features and present the results to users who are interested in Julia. Users who wish to play with the prototype GPU code for simulation can find the file in the test folder, but we warn users that the CuArrays package will not install if the proper GPU machinery is not detected. ","category":"page"},{"location":"#Demonstration","page":"Home","title":"Demonstration","text":"","category":"section"},{"location":"#Example-Data","page":"Home","title":"Example Data","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"We can use the OpenMendel package SnpArrays.jl to both read in and write out PLINK formatted files. Available in the data directory under the Example_Data section of this package, we use the file \"EUR_SUBSET\" for the demonstration how to simulate phenotypic traits on PLINK formatted data. For convenience we use the common assumption that the residual covariance among two relatives can be captured by the additive genetic variance times twice the kinship coefficient.","category":"page"},{"location":"","page":"Home","title":"Home","text":"In each simulation model, the user can specify parameters, along with the number of repetitions for each simulation model as desired. By default, the simulation will return the result of a single simulation.","category":"page"},{"location":"#Double-check-that-you-are-using-Julia-version-1.0-or-higher-by-checking-the-machine-information","page":"Home","title":"Double check that you are using Julia version 1.0 or higher by checking the machine information","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"versioninfo()","category":"page"},{"location":"","page":"Home","title":"Home","text":"Julia Version 1.4.0\nCommit b8e9a9ecc6 (2020-03-21 16:36 UTC)\nPlatform Info:\n  OS: macOS (x86_64-apple-darwin18.6.0)\n  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz\n  WORD_SIZE: 64\n  LIBM: libopenlibm\n  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)","category":"page"},{"location":"","page":"Home","title":"Home","text":"using Random, Plots, DataFrames, LinearAlgebra \nusing SnpArrays, TraitSimulation, GLM, StatsBase, OrdinalMultinomialModels\nRandom.seed!(1234);","category":"page"},{"location":"#Reading-genotype-data-using-SnpArrays","page":"Home","title":"Reading genotype data using SnpArrays","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"First use SnpArrays.jl to read in the genotype data. We use PLINK formatted data with the same prefixes for the .bim, .fam, .bed files.","category":"page"},{"location":"","page":"Home","title":"Home","text":"SnpArrays is a very useful utility and can do a lot more than just read in the data. More information about all the functionality of SnpArrays can be found at: https://openmendel.github.io/SnpArrays.jl/latest/","category":"page"},{"location":"","page":"Home","title":"Home","text":"As missing genotypes are often due to problems making the calls, the called genotypes at a marker with too much missing genotypes are potentially unreliable. By default, SnpArrays filters to keep only the genotypes with success rates greater than 0.98 and the minimum minor allele frequency to be 0.01. If the user wishes to change the stringency, change the number given in filter according to SnpArrays.","category":"page"},{"location":"","page":"Home","title":"Home","text":"filename = \"EUR_subset\"\nEUR = SnpArray(SnpArrays.datadir(filename * \".bed\"));","category":"page"},{"location":"","page":"Home","title":"Home","text":"rowmask, colmask =  SnpArrays.filter(EUR)\nminor_allele_frequency = maf(EUR);\npeople, snps = size(EUR)","category":"page"},{"location":"","page":"Home","title":"Home","text":"(379, 54051)","category":"page"},{"location":"","page":"Home","title":"Home","text":"EUR_data = SnpData(SnpArrays.datadir(filename));","category":"page"},{"location":"","page":"Home","title":"Home","text":"Here we will use identify by name, which locus to include, first subset the names of all the loci into a vector called snpid  and then call the following command to store our design matrix for the model that includes sex and locus of choice.","category":"page"},{"location":"","page":"Home","title":"Home","text":"bimfile = EUR_data.snp_info # store the snp_info with the snp names\n\nsnpid  = bimfile[!, :snpid] # store the snp names in the snpid vector\n\ncausal_snp_index = findall(x -> x == \"rs150018646\", snpid) # find the index of the snp of interest by snpid","category":"page"},{"location":"","page":"Home","title":"Home","text":"1-element Array{Int64,1}:\n 82","category":"page"},{"location":"","page":"Home","title":"Home","text":"Additionally, we will control for sex, with females as the baseline group, sex = 0.0. We want to find the index of this causal locus in the snp_definition (.bim) file and then subset that locus from the genetic marker data above. Make note of julia's ternary operator '?' which allows us to make this conversion efficiently!","category":"page"},{"location":"","page":"Home","title":"Home","text":"Using SnpArrays.jl we can then use the convert and @view commands to get the appropriate conversion from SnpArray to a computable vector of Float64. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"locus = convert(Vector{Float64}, @view(EUR[:, causal_snp_index[1]]))\nfamfile = EUR_data.person_info\nsex = map(x -> strip(x) == \"F\" ? 0.0 : 1.0, famfile[!, :sex])\nintercept = ones(length(sex))\nX_covar = [intercept sex]\nX = [intercept sex locus]","category":"page"},{"location":"","page":"Home","title":"Home","text":"379Ã—3 Array{Float64,2}:\n 1.0  1.0  2.0\n 1.0  1.0  1.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n â‹®         \n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  1.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0\n 1.0  1.0  2.0","category":"page"},{"location":"#Example-1:-GLM-Trait","page":"Home","title":"Example 1: GLM Trait","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"In this example we first demonstrate how to use the GLM.jl package to simulate a trait from unrelated individuals.  We specify the fixed effects and the phenotype distribution, and output the simulation results.","category":"page"},{"location":"","page":"Home","title":"Home","text":"Y sim Poisson(mu = g^-1(XB))","category":"page"},{"location":"","page":"Home","title":"Home","text":"A more thorough application of this GLM TraitSimulation applied to an Iterative Hard Thresholding problem under MendelIHT can be found here:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Pages = [\n    \"examples/testing_MendelIHT_glm.md\"\n]\nDepth = 2","category":"page"},{"location":"#GLM-Traits-from-Unrelated-Individuals","page":"Home","title":"GLM Traits from Unrelated Individuals","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"$ Y{n \\times 1} \\sim Poisson(\\mu{n \\times 1} = X\\beta)$","category":"page"},{"location":"","page":"Home","title":"Home","text":"Î² = [1; 0.2; 0.5]\ndist = Poisson()\nlink = LogLink()\nGLMmodel = GLMTrait(X, Î², dist, link)","category":"page"},{"location":"","page":"Home","title":"Home","text":"Generalized Linear Model\n  * response distribution: Poisson\n  * link function: LogLink\n  * sample size: 379  * fixed effects: 3","category":"page"},{"location":"","page":"Home","title":"Home","text":"Simulated_GLM_Traits = simulate(GLMmodel)","category":"page"},{"location":"","page":"Home","title":"Home","text":"379-element Array{Int64,1}:\n  4\n  8\n  7\n 10\n 10\n 11\n  7\n  9\n  8\n  9\n  6\n  9\n 10\n  â‹®\n  7\n 11\n  5\n 11\n  9\n  7\n  5\n  8\n  4\n 12\n 16\n 10","category":"page"},{"location":"#Example-2:-VCM-Trait","page":"Home","title":"Example 2: VCM Trait","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"In this example we show how to generate data so that the related individuals have correlated trait values even after we account for the effect of a snp, a combination of snps or other fixed effects. We simulate data under a variance component model so that we can model residual dependency among individuals. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"We note models with additional variance components can also be specified, as long as they are sensible (positive semi definite).","category":"page"},{"location":"#VCMTrait-UK-Biobank-Demo:","page":"Home","title":"VCMTrait UK Biobank Demo:","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"For the univariate and bivariate Variance Component Model, we include in the Real Data Examples section our step-by-step procedure of simultaneously simulating and fitting to get the genetic power. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"Y sim N(mu 4* 2GRM + 2*I_n)","category":"page"},{"location":"","page":"Home","title":"Home","text":"Y sim N(mathbfmu sigma_a otimes 2GRM + sigma_e otimes I_n)","category":"page"},{"location":"","page":"Home","title":"Home","text":"where we can calculate the estimated empirical kinship matrix 2*hatPhi_GRM using SnpArrays.jl. Due to privacy of the data, we include only the html file for viewing. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"Pages = [\n    \"examples/ukbiobank_vcm_power.md\"\n]\nDepth = 2","category":"page"},{"location":"","page":"Home","title":"Home","text":"We include some other variance component model examples for family data as follows:","category":"page"},{"location":"#Rare-Variant-VCM-Related-Individuals","page":"Home","title":"Rare Variant VCM Related Individuals","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"In this example we show how to generate data so that the related individuals have correlated trait values even after we account for the effect of a snp, a combination of snps or other fixed effects. We simulate data under a linear mixed model so that we can model residual dependency among individuals. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"Y sim textNormal(mathbfmu_n times 1 = Xbeta Sigma_n times n = sigma_A times 2hatPhi_GRM + sigma_E times I_n)","category":"page"},{"location":"","page":"Home","title":"Home","text":"This example is meant to simulate data in a scenario in which a number of rare mutations in a single gene can change a trait value. We model the residual variation among relatives with the additive genetic variance component and we include 20 simulated rare variants in the mean portion of the model, defined as loci with minor allele frequencies greater than 0.002 but less than 0.02.","category":"page"},{"location":"","page":"Home","title":"Home","text":"Specifically we are generating a single normal trait controlling for family structure with residual heritabiity of 67%, and effect sizes for the variants generated as a function of the minor allele frequencies. The rarer the variant the greater its effect size.","category":"page"},{"location":"","page":"Home","title":"Home","text":"In practice rare variants have smaller minor allele frequencies, but we are limited in this tutorial by the relatively small size of the data set. Note also that our modeling these effects as part of the mean is not meant to imply that the best way to detect them would be a standard association analysis. Instead we recommend a burden or SKAT test.","category":"page"},{"location":"","page":"Home","title":"Home","text":"GRM = grm(EUR, minmaf = 0.05);","category":"page"},{"location":"#Genotype-Simulation:","page":"Home","title":"Genotype Simulation:","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Say our study population has a sample size of n people and we are interested in studying the effect of the causal snp with a predetermined minor allele frequency. We use the minor allele frequency of the causal variant to simulate the SnpArray under Hardy Weinberg Equillibrium (HWE), using the snparray_simulation function as follows:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Genotype Plink/SnpArray\nA1,A1 0x00\nmissing 0x01\nA1,A2 0x02\nA2,A2 0x03","category":"page"},{"location":"","page":"Home","title":"Home","text":"Given a vector of minor allele frequencies, specify maf = [0.2, 0.25, 0.3], for each specified allele it will simulate a SnpArray under HWE and ouput them together. This function samples from the genotype vector under HWE and returns the compressed binary format under SnpArrays.","category":"page"},{"location":"","page":"Home","title":"Home","text":"maf_rare = rand([0.002, 0.004, 0.008, 0.01, 0.012, 0.015, 0.02], 20)\nrare_snps = snparray_simulation(maf_rare, 379)","category":"page"},{"location":"","page":"Home","title":"Home","text":"379Ã—20 SnpArray:\n 0x00  0x00  0x00  0x00  0x00  0x00  â€¦  0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00  â€¦  0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00  â€¦  0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n    â‹®                             â‹®  â‹±           â‹®                    \n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00  â€¦  0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x02  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x02  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00  â€¦  0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x02  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x00  0x00  0x00  0x00  0x00  0x00\n 0x00  0x00  0x00  0x00  0x00  0x00     0x02  0x00  0x00  0x00  0x00  0x00","category":"page"},{"location":"","page":"Home","title":"Home","text":"Î²_covar = [1.0; 0.6]\nI_n = Matrix{Float64}(I, size(GRM))\nvc = @vc [0.01][:,:] âŠ— (GRM + I_n) + [0.9][:, :] âŠ— I_n\nsigma, v = vcobjtuple(vc)\nrare_20_snp_model = VCMTrait(X_covar, Î²_covar[:, :], rare_snps, maf_rare[:, :], [sigma...], [v...])","category":"page"},{"location":"","page":"Home","title":"Home","text":"Variance Component Model\n  * number of traits: 1\n  * number of variance components: 2\n  * sample size: 379","category":"page"},{"location":"","page":"Home","title":"Home","text":"Rare_SNP_Trait = simulate(rare_20_snp_model)","category":"page"},{"location":"","page":"Home","title":"Home","text":"379Ã—1 Array{Float64,2}:\n 0.8088744352540489\n 1.6701367617702745\n 1.8455248160731537\n 0.8508388343337788\n 0.9712931279208524\n 0.6887865756651992\n 2.9436570129936905\n 1.808457300723279\n 1.1133732340741647\n 1.0798778322598372\n 0.8779238588613009\n 2.338615502643955\n 3.392334665830318\n â‹®\n 3.654119285594163\n 1.4716923062049176\n 1.7276952768730827\n 1.3970739033596746\n 1.4755462272829616\n 2.981506123686394\n 1.6977542002518275\n 3.09839068211269\n 3.225180822413689\n 2.7490032145653913\n 3.8705735842017255\n 2.7194612533505413","category":"page"},{"location":"#Multiple-Traits,-Multiple-Variance-Components?-Easy.","page":"Home","title":"Multiple Traits, Multiple Variance Components? Easy.","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"This example extends the variance component model in the previous example to demo how to efficiently account for any number of other random effects, in addition to the additive genetic and environmental variance components. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"Y ~ MatrixNormal(M = XB, Omega = sum_k=1^m Sigma_k otimes V_k)","category":"page"},{"location":"","page":"Home","title":"Home","text":"We note that this form can also accompany more than 2 variance components.","category":"page"},{"location":"","page":"Home","title":"Home","text":"I encourage for those interested, to look at this example where we demonstrate the simlation of d = 2 traits with m = 10 variance components, and benchmark it against the available method using the MatrixNormal distribution in Julia package, Distributions.jl.","category":"page"},{"location":"#Example-3:-Ordered-Multinomial-Trait","page":"Home","title":"Example 3: Ordered Multinomial Trait","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Now we demonstrate on the OrderedMultinomialTrait model object in TraitSimulation.jl.","category":"page"},{"location":"#Ordered-Multinomial-Trait","page":"Home","title":"Ordered Multinomial Trait","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Recall that this phenotype is special, in that the OrdinalMultinomialModels package provides Julia utilities to fit ordered multinomial models, including proportional odds model and ordered Probit model as special cases. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"For this model, we have intercept parameters theta_1 le cdots le theta_J-1 enforce the order between the categories and boldsymbolbeta reflects the effects of the linear predictors under the proportional hazards model. The effect sizes can be interpreted as the expected change of the response variable on an ordered log-odds scale for each unit increase in the predictor. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"Î¸ = [1.0, 1.2, 1.4]\nOrdinal_Model = OrderedMultinomialTrait(X, Î², Î¸, LogitLink())","category":"page"},{"location":"","page":"Home","title":"Home","text":"Ordinal Multinomial Model\n  * number of fixed effects: 3\n  * number of ordinal multinomial outcome categories: 4\n  * link function: LogitLink\n  * sample size: 379","category":"page"},{"location":"","page":"Home","title":"Home","text":"Ordinal_Trait = simulate(Ordinal_Model)","category":"page"},{"location":"","page":"Home","title":"Home","text":"379-element Array{Int64,1}:\n 2\n 4\n 4\n 1\n 1\n 1\n 1\n 4\n 1\n 4\n 4\n 3\n 4\n â‹®\n 4\n 4\n 2\n 1\n 1\n 4\n 1\n 4\n 4\n 4\n 4\n 4","category":"page"},{"location":"#Simulate-Ordered-Multinomial-Logistic","page":"Home","title":"Simulate Ordered Multinomial Logistic","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Specific to the Ordered Multinomial Logistic model is the option to transform the multinomial outcome (i.e 1, 2, 3, 4) into a binary outcome for logistic regression. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"Although by default is the multinomial simulation above, the user can simulate from the transformed logistic outcome for example by specifying arguments: Logistic = true and threshold = 2 the value to use as a cutoff for identifying cases and controls. (i.e if y > 2 => y_logit == 1). We note if you specify Logistic = true and do not provide a threshold value, the function will throw an error to remind you to specify one.","category":"page"},{"location":"","page":"Home","title":"Home","text":"Logistic_Trait = simulate(Ordinal_Model, Logistic = true, threshold = 2)","category":"page"},{"location":"","page":"Home","title":"Home","text":"379-element Array{Int64,1}:\n 1\n 1\n 1\n 1\n 1\n 1\n 0\n 0\n 1\n 1\n 1\n 0\n 1\n â‹®\n 1\n 1\n 0\n 1\n 1\n 1\n 0\n 0\n 1\n 1\n 1\n 1","category":"page"},{"location":"#Example-4:-GLMM-Trait-Simulation","page":"Home","title":"Example 4: GLMM Trait Simulation","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Next, we demonstrate how to simulate a Poisson Trait, after controlling for family structure.  ","category":"page"},{"location":"","page":"Home","title":"Home","text":"dist = Poisson()\nlink = LogLink()\n\nvc = @vc [0.01][:,:] âŠ— (GRM + I_n) + [0.9][:, :] âŠ— I_n\nGLMMmodel = GLMMTrait(X, Î²[:,:], vc, dist, link)","category":"page"},{"location":"","page":"Home","title":"Home","text":"Generalized Linear Mixed Model\n  * response distribution: Poisson\n  * link function: LogLink\n  * number of variance components: 2\n  * sample size: 379","category":"page"},{"location":"","page":"Home","title":"Home","text":"Y = simulate(GLMMmodel)","category":"page"},{"location":"","page":"Home","title":"Home","text":"379Ã—1 Array{Int64,2}:\n 31\n  3\n  5\n  4\n  9\n 10\n  3\n 10\n 17\n  6\n 16\n 24\n 16\n  â‹®\n 18\n  3\n  4\n  2\n 28\n  5\n  1\n  6\n 32\n  0\n  9\n 13","category":"page"},{"location":"#Power-Calculation","page":"Home","title":"Power Calculation","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"(Image: png)","category":"page"},{"location":"","page":"Home","title":"Home","text":"Now we demonstrate the use of the simulations to generates data sets allowing researchers to accurately check the validity of programs and to calculate power for their proposed studies. For these examples we will demo the full power pipeline contatined within the OpenMendel environment.","category":"page"},{"location":"","page":"Home","title":"Home","text":"We illustrate this example in three digestable steps as shown in the figure: ","category":"page"},{"location":"","page":"Home","title":"Home","text":"The first by simulating genotypes and covariate values representative for our study population.\nCarry over the simulated design matrix from (1) to create the OrderedMultinomialTrait model object.\nSimulate off the OrderedMultinomialTrait model object created in (2) and run the power analyses for the desired significance level.","category":"page"},{"location":"","page":"Home","title":"Home","text":"Each column of this matrix represents each of the detected effect sizes, and each row of this matrix represents each simulation for that effect size. The user feeds into the function the number of simulations, the vector of effect sizes, the TraitSimulation.jl model object, and the random seed. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"mathbfY_n times 1 = mathbfXmathbfbeta + mathbfG_s gamma + mathbfg + mathbfepsilon","category":"page"},{"location":"","page":"Home","title":"Home","text":"mathbfg sim N(mathbf0 sigma_A times mathbfPhi)","category":"page"},{"location":"","page":"Home","title":"Home","text":"mathbfepsilon sim N(mathbf0 sigma_E times mathbfI_n)","category":"page"},{"location":"","page":"Home","title":"Home","text":"Here sigma_A and Sigma_A are the additive genetic variance and matrix, sigma_E and Sigma_E are the environmental variance and matrix, mathbfPhi is the kinship matrix, and mathbfI_n is the n times n identity matrix.","category":"page"},{"location":"","page":"Home","title":"Home","text":"We illustrate the example on simulated data under the Ordered Multinomial Model and define the study power in the following way. For each simulated trait, we perform a likelihood ratio test of the above hypothesis test and reject the null when the p-value falls below prespecified significance level. In our example, we define the power for the model is estimated as the proportion of tests rejecting the null at significance level Î± = 5Ã—10âˆ’8.","category":"page"},{"location":"","page":"Home","title":"Home","text":"textbfH_0 gamma = 0  hspace025intextversushspace025in textbfH_A gamma neq 0","category":"page"},{"location":"","page":"Home","title":"Home","text":"For GLMTrait objects, the realistic_power_simulation function makes the appropriate calls to the GLM.jl package to get the simulation p-values obtained from testing the significance of the causal locus using the Wald Test by default. However since the GLM.jl package has its limitations, we include additional power utilities that make the appropriate function calls to the OrdinalMultinomialModels to get the p-value obtained from testing the significance of the causal locus. Variance Component Models for VCMTrait objects can be fit using VarianceComponentModels.jl, and we demo for UK Biobank data in the docs. ","category":"page"},{"location":"#UKBiobank-Power-Pipeline","page":"Home","title":"UKBiobank Power Pipeline","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Pages = [\n    \"examples/ukbiobank_vcm_power.md\",\n    \"examples/ukbiobank_ordered_multinomial_power.md\"\n]\nDepth = 2","category":"page"},{"location":"","page":"Home","title":"Home","text":"For each effect size in gamma_s in each column we have the p-values obtained from testing the significance of the causal locus nsim = 1000 times under the desired model, and the randomseed = 1234.","category":"page"},{"location":"#Try-it-Yourself-Exercises:","page":"Home","title":"Try it Yourself Exercises:","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"With the remaining time we encourage you to play with the simulation models above. ","category":"page"},{"location":"","page":"Home","title":"Home","text":"(1) If you change the significance level, Î±, what would happen to the plot of the power? How would the plot change with a smaller value of Î± ? ","category":"page"},{"location":"","page":"Home","title":"Home","text":"(2) What if you change the sample size to be smaller than n = 5000? How does the power look as a function of the effect size when n = 2500? ","category":"page"},{"location":"","page":"Home","title":"Home","text":"(3) What if we changed the minor allele frequency of the SNP of interest from maf = 0.2 to say maf = 0.3 ?","category":"page"},{"location":"#Citations:","page":"Home","title":"Citations:","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"[1] Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.`","category":"page"},{"location":"","page":"Home","title":"Home","text":"[2] OPENMENDEL: a cooperative programming project for statistical genetics. Hum Genet. 2019 Mar 26. doi: 10.1007/s00439-019-02001-z.","category":"page"},{"location":"","page":"Home","title":"Home","text":"[3] German, CA, Sinsheimer, JS, Klimentidis, YC, Zhou, H, Zhou, JJ. Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale. Genetic Epidemiology. 2019; 1â€“ 13. https://doi.org/10.1002/gepi.22276","category":"page"}]
}
