<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Testing Methods Â· TraitSimulation.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">TraitSimulation.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li class="is-active"><a class="tocitem" href>Testing Methods</a><ul class="internal"><li><a class="tocitem" href="#Simulate-the-GLM-Trait"><span>Simulate the GLM Trait</span></a></li><li class="toplevel"><a class="tocitem" href="#Run-Mendel-IHT"><span>Run Mendel IHT</span></a></li><li><a class="tocitem" href="#Citations:"><span>Citations:</span></a></li></ul></li><li><span class="tocitem">Real Data Examples</span><ul><li><a class="tocitem" href="../ukbiobank_vcm_power/">Trait Simulation - Variance Component Model Power (UKBiobank)</a></li><li><a class="tocitem" href="../ukbiobank_ordered_multinomial_power/">Trait Simulation - Ordinal Multinomial Power</a></li></ul></li><li><a class="tocitem" href="../GPU/">Exploring GPU Potential</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Testing Methods</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Testing Methods</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/TraitSimulation.jl/blob/master/docs/src/examples/testing_MendelIHT_glm.md" title="Edit on GitHub"><span class="docs-icon fab">ï‚›</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Trait-Simulation:-Prototyping-Analysis-Methods-(MendelIHT)"><a class="docs-heading-anchor" href="#Trait-Simulation:-Prototyping-Analysis-Methods-(MendelIHT)">Trait Simulation: Prototyping Analysis Methods (MendelIHT)</a><a id="Trait-Simulation:-Prototyping-Analysis-Methods-(MendelIHT)-1"></a><a class="docs-heading-anchor-permalink" href="#Trait-Simulation:-Prototyping-Analysis-Methods-(MendelIHT)" title="Permalink"></a></h1><p>Authors: Sarah Ji, Benjamin Chu, Janet Sinsheimer, Kenneth Lange, Hua Zhou</p><p>This software package, TraitSimuliation.jl addresses the need for simulated trait data in genetic analyses.  This package generates data sets that will allow researchers to accurately check the validity of programs and to calculate power for their proposed studies. </p><p>In this notebook we show how to the <code>TraitSimulation.jl</code> package can be used to validate the results of the newest OpenMendel analysis package, <code>MendelIHT.jl</code>.</p><p>First we will generate the desired non-genetic covariates for our examples, sex and age from the <code>GLM.jl</code> package. Then we will make the appropriate calls to OpenMendel packages SnpArrays.jl, MendelIHT.jl. and TraitSimulation.jl to construct the desired genetic model, simulate from it and test the iterative hard thresholding model.</p><p>For the following three distributions, we simulate traits and validate the results using MendelIHT.</p><ol><li>Normal Trait</li><li>Poisson Trait</li><li>Negative Binomial Trait</li></ol><p>Double check that you are using Julia version 1.0 or higher by checking the machine information</p><pre><code class="language-julia">versioninfo()</code></pre><pre><code class="language-none">Julia Version 1.2.0
Commit c6da87ff4b (2019-08-20 00:03 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.6.0)
  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)</code></pre><pre><code class="language-julia">using Random, DataFrames, LinearAlgebra
using SnpArrays, TraitSimulation, GLM, StatsBase, MendelIHT
using Revise
revise()

Random.seed!(123);</code></pre><p>Simulate Non-Genetic Covariates</p><pre><code class="language-julia">n = 5000
p = 10000
s = 5

# simulate non genetic covariates
intercept = ones(n, 1) 
sex = rand(Bernoulli(0.51), n)
age = zscore(rand(Normal(45, 8), n))
non_genetic_cov =  [sex age]
Î²_non_gen = rand(size(non_genetic_cov, 2))</code></pre><pre><code class="language-none">2-element Array{Float64,1}:
 0.7373240898122129
 0.4314238505987109</code></pre><p>Simulate Genetic Covariates </p><p>First we will generate the entire snparray of p = 10,000 snps for n = 1000 people. Then for s = 10 causal snps, we will convert them to Float64 integers for simulation.</p><pre><code class="language-julia"># generate entire snparray
x = simulate_random_snparray(n, p, &quot;tmp.bed&quot;)
xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true); 
k = 5
# generate snparray with only causal snps
causal_idx = rand(1:10000, k) # these are the true indices
causal_effect_sizes = rand(k) # these are the true effect sizes

x_causal = convert(Matrix{Float64}, @view(x[:, causal_idx]), center = true, scale = true);</code></pre><h2 id="Simulate-the-GLM-Trait"><a class="docs-heading-anchor" href="#Simulate-the-GLM-Trait">Simulate the GLM Trait</a><a id="Simulate-the-GLM-Trait-1"></a><a class="docs-heading-anchor-permalink" href="#Simulate-the-GLM-Trait" title="Permalink"></a></h2><p>Use TraitSimulation.jl to construct the desired model and simulate from it. For the following three distributions, we simulate traits and validate the results using MendelIHT.</p><ol><li>Normal Trait</li><li>Poisson Trait</li><li>Negative Binomial Trait</li></ol><h3 id="Testing-Normal-Trait"><a class="docs-heading-anchor" href="#Testing-Normal-Trait">Testing Normal Trait</a><a id="Testing-Normal-Trait-1"></a><a class="docs-heading-anchor-permalink" href="#Testing-Normal-Trait" title="Permalink"></a></h3><pre><code class="language-julia">distribution = Normal()
link = IdentityLink()
# revise()
GLM_model = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, distribution, link)
y = simulate(GLM_model)</code></pre><pre><code class="language-none">5000-element Array{Float64,1}:
  0.9185868876881607    
 -0.9518262679882143    
 -0.33841827510588013   
  2.049817083313287     
  1.955913151958258     
  0.00024397836600864933
 -0.6593742370456934    
 -0.28351854873874294   
  0.3185448435991667    
  0.30756420729346534   
 -0.4628928677570747    
  1.292210040257612     
  0.06891346416538825   
  â‹®                     
 -1.0982567760970272    
 -0.016491104287002845  
  0.9345002196941151    
  0.14027327315163013   
 -2.20745242227931      
  1.8222300720529194    
  2.9752072523810957    
  0.4504020749944919    
 -1.7701081883560879    
 -0.3661636604161142    
  2.5102925366869027    
  0.9083339384177506</code></pre><h1 id="Run-Mendel-IHT"><a class="docs-heading-anchor" href="#Run-Mendel-IHT">Run Mendel IHT</a><a id="Run-Mendel-IHT-1"></a><a class="docs-heading-anchor-permalink" href="#Run-Mendel-IHT" title="Permalink"></a></h1><pre><code class="language-julia">s_guess = 7
@show Î²_non_gen
@show truth = causal_idx 
@show causal_effect_sizes
result = L0_reg(x, xbm, non_genetic_cov, y, 50, s_guess, Normal(), IdentityLink(), debias = false, max_iter = 500, verbose= false)</code></pre><pre><code class="language-none">Î²_non_gen = [0.7373240898122129, 0.4314238505987109]
truth = causal_idx = [184, 4492, 5382, 161, 2773]
causal_effect_sizes = [0.6877212499518375, 0.7610202033321085, 0.6111539989474444, 0.7062198720454749, 0.06323470301651035]






IHT estimated 5 nonzero SNP predictors and 2 non-genetic predictors.

Compute time (sec):     2.9930989742279053
Final loglikelihood:    -7115.458807213961
Iterations:             10

Selected genetic predictors:
5Ã—2 DataFrame
â”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚
â”‚     â”‚ [90mInt64[39m    â”‚ [90mFloat64[39m     â”‚
â”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ 1   â”‚ 161      â”‚ 0.702183    â”‚
â”‚ 2   â”‚ 184      â”‚ 0.68639     â”‚
â”‚ 3   â”‚ 4492     â”‚ 0.75284     â”‚
â”‚ 4   â”‚ 5382     â”‚ 0.595549    â”‚
â”‚ 5   â”‚ 5867     â”‚ 0.0581857   â”‚

Selected nongenetic predictors:
2Ã—2 DataFrame
â”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚
â”‚     â”‚ [90mInt64[39m    â”‚ [90mFloat64[39m     â”‚
â”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ 1   â”‚ 1        â”‚ 0.715045    â”‚
â”‚ 2   â”‚ 2        â”‚ 0.428766    â”‚</code></pre><h3 id="Testing-Poisson-Trait"><a class="docs-heading-anchor" href="#Testing-Poisson-Trait">Testing Poisson Trait</a><a id="Testing-Poisson-Trait-1"></a><a class="docs-heading-anchor-permalink" href="#Testing-Poisson-Trait" title="Permalink"></a></h3><pre><code class="language-julia">distribution2 = Poisson()
link2 = LogLink()

non_genetic_cov =  [sex age]
Î²_non_gen = [0.1, 0.002]</code></pre><pre><code class="language-none">2-element Array{Float64,1}:
 0.1  
 0.002</code></pre><pre><code class="language-julia">GLM_model2 = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, Poisson, link2)</code></pre><pre><code class="language-none">Generalized Linear Model
  * response distribution: UnionAll
  * link function: LogLink
  * sample size: 5000</code></pre><pre><code class="language-julia">x = simulate_random_snparray(n, p, &quot;tmp.bed&quot;)
xbm = SnpBitMatrix{Float64}(x, model=ADDITIVE_MODEL, center=true, scale=true);
k = 5
causal_idx = rand(1:10000, k) # these are the true indices
causal_effect_sizes = rand(k)
x_causal = convert(Matrix{Float64}, @view(x[:, causal_idx]), center = true, scale = true);</code></pre><pre><code class="language-julia"># construct the model 
GLM_model_p2 = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, Poisson(), link2)</code></pre><pre><code class="language-none">Generalized Linear Model
  * response distribution: Poisson
  * link function: LogLink
  * sample size: 5000</code></pre><pre><code class="language-julia"># simulate the trait
y_poisson2 = simulate(GLM_model_p2)</code></pre><pre><code class="language-none">5000-element Array{Int64,1}:
 0
 3
 0
 0
 2
 0
 0
 0
 0
 7
 1
 0
 2
 â‹®
 6
 1
 2
 4
 4
 2
 0
 0
 1
 3
 1
 0</code></pre><pre><code class="language-julia">s_guess = 7
result2 = L0_reg(x, xbm, non_genetic_cov, Float64.(y_poisson2), 50, s_guess, Poisson(), LogLink(), debias = false, max_iter = 500, verbose= false)</code></pre><pre><code class="language-none">IHT estimated 6 nonzero SNP predictors and 1 non-genetic predictors.

Compute time (sec):     8.748939990997314
Final loglikelihood:    -6742.100029722215
Iterations:             32

Selected genetic predictors:
6Ã—2 DataFrame
â”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚
â”‚     â”‚ [90mInt64[39m    â”‚ [90mFloat64[39m     â”‚
â”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ 1   â”‚ 400      â”‚ 0.405941    â”‚
â”‚ 2   â”‚ 2943     â”‚ 0.0423735   â”‚
â”‚ 3   â”‚ 3749     â”‚ 0.036553    â”‚
â”‚ 4   â”‚ 4656     â”‚ 0.53714     â”‚
â”‚ 5   â”‚ 7563     â”‚ 0.722796    â”‚
â”‚ 6   â”‚ 8016     â”‚ 0.239444    â”‚

Selected nongenetic predictors:
1Ã—2 DataFrame
â”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚
â”‚     â”‚ [90mInt64[39m    â”‚ [90mFloat64[39m     â”‚
â”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ 1   â”‚ 1        â”‚ 0.0952795   â”‚</code></pre><h3 id="Testing-Negative-Binomial-Trait"><a class="docs-heading-anchor" href="#Testing-Negative-Binomial-Trait">Testing Negative Binomial Trait</a><a id="Testing-Negative-Binomial-Trait-1"></a><a class="docs-heading-anchor-permalink" href="#Testing-Negative-Binomial-Trait" title="Permalink"></a></h3><pre><code class="language-julia"># simulate the trait
distribution = NegativeBinomial()
link = LogLink()

GLM_model = GLMTrait(non_genetic_cov, Î²_non_gen, x_causal, causal_effect_sizes, distribution, link)
y_negbinomial = simulate(GLM_model)</code></pre><pre><code class="language-none">5000-element Array{Int64,1}:
  1
  1
  0
  0
  2
  1
  1
  1
  0
  1
  0
  1
  0
  â‹
 15
  2
  3
  2
 13
  1
  0
  0
  1
  1
  0
  0</code></pre><pre><code class="language-julia"># run IHT
s_guess = 7
@show Î²_non_gen
@show truth = causal_idx 
@show causal_effect_sizes
result = L0_reg(x, xbm, non_genetic_cov, Float64.(y_negbinomial), 50, s_guess, NegativeBinomial(), link, debias = false, use_maf = false, max_iter = 500, verbose= false)</code></pre><pre><code class="language-none">Î²_non_gen = [0.1, 0.002]
truth = causal_idx = [4656, 9394, 400, 8016, 7563]
causal_effect_sizes = [0.5264504200595785, 0.008085205500707904, 0.4110283000164554, 0.22887974718537518, 0.7299946760225595]






IHT estimated 6 nonzero SNP predictors and 1 non-genetic predictors.

Compute time (sec):     4.818864107131958
Final loglikelihood:    -7579.647514751706
Iterations:             17

Selected genetic predictors:
6Ã—2 DataFrame
â”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚
â”‚     â”‚ [90mInt64[39m    â”‚ [90mFloat64[39m     â”‚
â”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ 1   â”‚ 400      â”‚ 0.418851    â”‚
â”‚ 2   â”‚ 1118     â”‚ 0.0727529   â”‚
â”‚ 3   â”‚ 3367     â”‚ -0.0822334  â”‚
â”‚ 4   â”‚ 4656     â”‚ 0.50584     â”‚
â”‚ 5   â”‚ 7563     â”‚ 0.737677    â”‚
â”‚ 6   â”‚ 8016     â”‚ 0.193797    â”‚

Selected nongenetic predictors:
1Ã—2 DataFrame
â”‚ Row â”‚ Position â”‚ Estimated_Î² â”‚
â”‚     â”‚ [90mInt64[39m    â”‚ [90mFloat64[39m     â”‚
â”œâ”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ 1   â”‚ 1        â”‚ 0.140459    â”‚</code></pre><h2 id="Citations:"><a class="docs-heading-anchor" href="#Citations:">Citations:</a><a id="Citations:-1"></a><a class="docs-heading-anchor-permalink" href="#Citations:" title="Permalink"></a></h2><p>[1] Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.`</p><p>[2] OPENMENDEL: a cooperative programming project for statistical genetics. <a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=OPENMENDEL">Hum Genet. 2019 Mar 26. doi: 10.1007/s00439-019-02001-z</a>.</p><p>[3] Benjamin B. Chu, Kevin L. Keys, Christopher A. German, Hua Zhou, Jin J. Zhou, Janet S. Sinsheimer, Kenneth Lange. Iterative Hard Thresholding in GWAS: Generalized Linear Models, Prior Weights, and Double Sparsity. bioRxiv doi:10.1101/697755</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">Â« Home</a><a class="docs-footer-nextpage" href="../ukbiobank_vcm_power/">Trait Simulation - Variance Component Model Power (UKBiobank) Â»</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 28 December 2020 18:36">Monday 28 December 2020</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
